<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>CBW_2021_CAN_M7 Single Cell RNA</title>
  <meta name="description" content="CBW_2021_CAN_M7 Single Cell RNA" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="CBW_2021_CAN_M7 Single Cell RNA" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="CBW_2021_CAN_M7 Single Cell RNA" />
  
  
  

<meta name="author" content="J. Javier Diaz-Mejia" />


<meta name="date" content="2021-06-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CBW_2021_CAN_M7</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="#load-raw-data-qc-and-normalization"><i class="fa fa-check"></i><b>1</b> Load raw data, QC and normalization </a><ul>
<li class="chapter" data-level="1.0.1" data-path="index.html"><a href="#define-environment-inputs-and-outdir"><i class="fa fa-check"></i><b>1.0.1</b> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</a></li>
<li class="chapter" data-level="1.0.2" data-path="index.html"><a href="#load-datasets"><i class="fa fa-check"></i><b>1.0.2</b> LOAD DATASETS</a></li>
<li class="chapter" data-level="1.0.3" data-path="index.html"><a href="#qc-datasets"><i class="fa fa-check"></i><b>1.0.3</b> QC DATASETS</a></li>
<li class="chapter" data-level="1.0.4" data-path="index.html"><a href="#normalize-datasets"><i class="fa fa-check"></i><b>1.0.4</b> NORMALIZE DATASETS</a></li>
<li class="chapter" data-level="1.0.5" data-path="index.html"><a href="#save-each-dataset-r_object"><i class="fa fa-check"></i><b>1.0.5</b> SAVE EACH DATASET R_OBJECT</a></li>
<li class="chapter" data-level="1.0.6" data-path="index.html"><a href="#obtain-computing-time"><i class="fa fa-check"></i><b>1.0.6</b> OBTAIN COMPUTING TIME</a></li>
<li class="chapter" data-level="1.0.7" data-path="index.html"><a href="#finish"><i class="fa fa-check"></i><b>1.0.7</b> FINISH</a></li>
<li class="divider"></li>
<li class="chapter" data-level="2" data-path="index.html"><a href="#integration"><i class="fa fa-check"></i><b>2</b> Integration </a><ul>
<li class="chapter" data-level="2.0.1" data-path="index.html"><a href="#define-environment-inputs-and-outdir-1"><i class="fa fa-check"></i><b>2.0.1</b> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</a></li>
<li class="chapter" data-level="2.0.2" data-path="index.html"><a href="#load-r-objects"><i class="fa fa-check"></i><b>2.0.2</b> LOAD R OBJECTS</a></li>
<li class="chapter" data-level="2.0.3" data-path="index.html"><a href="#integrate-datasets"><i class="fa fa-check"></i><b>2.0.3</b> INTEGRATE DATASETS</a></li>
<li class="chapter" data-level="2.0.4" data-path="index.html"><a href="#integrate-datasets-1"><i class="fa fa-check"></i><b>2.0.4</b> INTEGRATE DATASETS</a></li>
<li class="chapter" data-level="2.0.5" data-path="index.html"><a href="#save-integrated-r_object"><i class="fa fa-check"></i><b>2.0.5</b> SAVE INTEGRATED R_OBJECT</a></li>
<li class="chapter" data-level="2.0.6" data-path="index.html"><a href="#obtain-computing-time-1"><i class="fa fa-check"></i><b>2.0.6</b> OBTAIN COMPUTING TIME</a></li>
<li class="chapter" data-level="2.0.7" data-path="index.html"><a href="#finish-1"><i class="fa fa-check"></i><b>2.0.6</b> FINISH</a></li>
<li class="divider"></li>
<li class="chapter" data-level="3" data-path="index.html"><a href="#dimension-reduction-clustering"><i class="fa fa-check"></i><b>1</b> Dimension reduction, cell clustering </a><ul>
<li class="chapter" data-level="3.0.15" data-path="index.html"><a href="#define-environment-inputs-and-outdir-2"><i class="fa fa-check"></i><b>1.0.15</b> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</a></li>
<li class="chapter" data-level="3.0.16" data-path="index.html"><a href="#load-r-object"><i class="fa fa-check"></i><b>1.0.16</b> LOAD R OBJECT</a></li>
<li class="chapter" data-level="3.0.17" data-path="index.html"><a href="#reduce-dimensions"><i class="fa fa-check"></i><b>1.0.17</b> REDUCE DIMENSIONS</a></li>
<li class="chapter" data-level="3.0.18" data-path="index.html"><a href="#cluster-cells"><i class="fa fa-check"></i><b>1.0.18</b> CLUSTER CELLS</a></li>
<li class="chapter" data-level="3.0.19" data-path="index.html"><a href="#generate-umap-plots"><i class="fa fa-check"></i><b>1.0.19</b> GENERATE UMAP PLOTS</a></li>
<li class="chapter" data-level="3.0.20" data-path="index.html"><a href="#generate-dot-plots"><i class="fa fa-check"></i><b>1.0.20</b> GENERATE DOT PLOTS</a></li>
<li class="chapter" data-level="3.0.21" data-path="index.html"><a href="#get-average-gene-expression-per-cluster"><i class="fa fa-check"></i><b>1.0.21</b> GET AVERAGE GENE EXPRESSION PER CLUSTER</a></li>
<li class="chapter" data-level="3.0.22" data-path="index.html"><a href="#save-integrated-r_object-1"><i class="fa fa-check"></i><b>1.0.22</b> SAVE INTEGRATED R_OBJECT</a></li>
<li class="chapter" data-level="3.0.23" data-path="index.html"><a href="#obtain-computing-time-2"><i class="fa fa-check"></i><b>1.0.23</b> OBTAIN COMPUTING TIME</a></li>
<li class="chapter" data-level="3.0.24" data-path="index.html"><a href="#finish-2"><i class="fa fa-check"></i><b>1.0.24</b> FINISH</a></li>
<li class="divider"></li>
<li class="chapter" data-level="4" data-path="index.html"><a href="#differential-gene-expression"><i class="fa fa-check"></i><b>1</b> Differential gene expression </a><ul>
<li class="chapter" data-level="4.0.25" data-path="index.html"><a href="#define-environment-inputs-and-outdir-3"><i class="fa fa-check"></i><b>1.0.25</b> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</a></li>
<li class="chapter" data-level="4.0.26" data-path="index.html"><a href="#load-r-object-1"><i class="fa fa-check"></i><b>1.0.26</b> LOAD R OBJECT</a></li>
<li class="chapter" data-level="4.0.27" data-path="index.html"><a href="#load-metadata"><i class="fa fa-check"></i><b>1.0.27</b> LOAD METADATA</a></li>
<li class="chapter" data-level="4.0.28" data-path="index.html"><a href="#find-dge-using-global-cell-clusters-dimensions"><i class="fa fa-check"></i><b>1.0.28</b> FIND DGE USING GLOBAL CELL CLUSTERS DIMENSIONS</a></li>
<li class="chapter" data-level="4.0.29" data-path="index.html"><a href="#find-dge-using-paired-metadata-classes"><i class="fa fa-check"></i><b>1.0.29</b> FIND DGE USING PAIRED METADATA CLASSES</a></li>
<li class="chapter" data-level="4.0.30" data-path="index.html"><a href="#obtain-computing-time-3"><i class="fa fa-check"></i><b>1.0.30</b> OBTAIN COMPUTING TIME</a></li>
<li class="chapter" data-level="4.0.31" data-path="index.html"><a href="#finish-3"><i class="fa fa-check"></i><b>1.0.31</b> FINISH</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">CBW_2021_CAN_M7 Single Cell RNA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">CBW_2021_CAN_M7 Single Cell RNA</h1>
<p class="author"><em>J. Javier Diaz-Mejia</em></p>
<p class="date" style="margin-top: 1.5em;"><em>2021-06-08</em></p>
</div>
<div id="down-the-rabbit-hole" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Down the rabbit-hole</h1>

<ul>
<li>Javier Diaz - <a href="mailto:javier.diazmejia@gmail.com" class="email">javier.diazmejia@gmail.com</a></li>
<li><p>Script based on: <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="uri">https://satijalab.org/seurat/articles/integration_introduction.html</a></p></li>
<li>GENERAL OVERVIEW OF THIS SCRIPT
<ol style="list-style-type: decimal">
<li>Loads scRNA-seq data from Cell Ranger (MTX files)</li>
<li>Generates QC plots for each dataset</li>
<li>Normalizes each dataset</li>
<li>Saves R objects with each normalized dataset</li>
</ol></li>
</ul>
<hr />
<div id="define-environment-inputs-and-outdir" class="section level3">
<h3><span class="header-section-number">1.0.1</span> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">oldw &lt;-<span class="st"> </span><span class="kw">getOption</span>(<span class="st">&quot;warn&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">options</span>( <span class="dt">warn =</span> <span class="dv">-1</span> )</a></code></pre></div>
<div id="required-libraries" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.1</span> Required libraries</h6>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(Seurat)   <span class="co"># (CRAN) main scRNA-seq analysis package</span></a></code></pre></div>
<pre><code>## Attaching SeuratObject</code></pre>
</div>
<div id="users-home-and-stopwatch-to-time-run" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.2</span> User’s home and stopwatch to time run</h6>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">UserHomeDirectory &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">print</span>(UserHomeDirectory)</a></code></pre></div>
<pre><code>## [1] &quot;/Users/jdiazmej&quot;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">StopWatchStart &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">StopWatchEnd   &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">StopWatchStart<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a></code></pre></div>
</div>
<div id="define-inputs" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.3</span> Define inputs</h6>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">### Select glioblastoma datasets from Richards_NatCancer_2021</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">DatasetIds &lt;-<span class="st"> </span><span class="kw">list</span>( <span class="co">## List of input IDs to process</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">  <span class="st">`</span><span class="dt">1</span><span class="st">`</span> =<span class="st"> &quot;G1003_A_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">  <span class="st">`</span><span class="dt">2</span><span class="st">`</span> =<span class="st"> &quot;G620_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">  <span class="st">`</span><span class="dt">3</span><span class="st">`</span> =<span class="st"> &quot;G910_A_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">  <span class="st">`</span><span class="dt">4</span><span class="st">`</span> =<span class="st"> &quot;G945_I_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">  <span class="st">`</span><span class="dt">5</span><span class="st">`</span> =<span class="st"> &quot;G946_I_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">  <span class="st">`</span><span class="dt">6</span><span class="st">`</span> =<span class="st"> &quot;G967_A_T&quot;</span>,</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">  <span class="st">`</span><span class="dt">7</span><span class="st">`</span> =<span class="st"> &quot;G983_A_T&quot;</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10">)</a>
<a class="sourceLine" id="cb7-11" data-line-number="11"></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">### /path_to/MTX_DIRECTORIES from 10X Cell Ranger (i.e. &#39;filtered_feature_bc_matrix&#39; directories</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">### with [features.tsv.gz, barcodes.tsv.gz and matrix.mtx.gz] files)</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14">PathToDatasets &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb7-15" data-line-number="15">  <span class="st">&quot;G1003_A_T&quot;</span> =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G1003_A_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-16" data-line-number="16">  <span class="st">&quot;G620_T&quot;</span>    =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G620_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-17" data-line-number="17">  <span class="st">&quot;G910_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G910_A_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-18" data-line-number="18">  <span class="st">&quot;G945_I_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G945_I_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-19" data-line-number="19">  <span class="st">&quot;G946_I_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G946_I_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-20" data-line-number="20">  <span class="st">&quot;G967_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G967_A_T/&quot;</span>,</a>
<a class="sourceLine" id="cb7-21" data-line-number="21">  <span class="st">&quot;G983_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G983_A_T/&quot;</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22">)</a></code></pre></div>
</div>
<div id="define-outputs" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.4</span> Define outputs</h6>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">### Outputs</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">PathForOutfiles &lt;-<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021&quot;</span> <span class="co">## /path_to/out_directory</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">PrefixOutfiles  &lt;-<span class="st"> &quot;Richards_NatCancer_2021&quot;</span>  <span class="co">## Prefix for outfiles</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4">PathForOutfiles &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathForOutfiles)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="kw">dir.create</span>(<span class="dt">path =</span> PathForOutfiles, <span class="dt">recursive =</span> T, <span class="dt">showWarnings =</span> F)</a></code></pre></div>
</div>
<div id="define-default-parameters" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.5</span> Define default parameters</h6>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">DefaultParameters &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">  </a>
<a class="sourceLine" id="cb9-3" data-line-number="3">  <span class="co">### Parameters for QC plots</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4">  <span class="dt">CellPropertiesToQC =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;mito.fraction&quot;</span>),</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">  </a>
<a class="sourceLine" id="cb9-6" data-line-number="6">  <span class="co">### Parameters for Seurat filters</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7">  <span class="dt">MinCells =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb9-8" data-line-number="8">  <span class="dt">MinGenes =</span> <span class="dv">50</span>,</a>
<a class="sourceLine" id="cb9-9" data-line-number="9">  <span class="dt">MaxReads =</span> <span class="dv">80000</span>,</a>
<a class="sourceLine" id="cb9-10" data-line-number="10">  <span class="dt">MaxMitoFraction =</span> <span class="fl">0.2</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11">)</a></code></pre></div>
</div>
<div id="report-r-sessioninfo" class="section level6">
<h6><span class="header-section-number">1.0.1.0.0.6</span> Report R sessionInfo</h6>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">OutfileRSessionInfo&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_SingleCell_1_QC_Normalization_RSessionInfo.txt&quot;</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw">writeLines</span>(<span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>()), OutfileRSessionInfo)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>())</a></code></pre></div>
<pre><code>##  [1] &quot;R version 4.0.2 (2020-06-22)&quot;                                                                                           
##  [2] &quot;Platform: x86_64-apple-darwin17.0 (64-bit)&quot;                                                                             
##  [3] &quot;Running under: macOS High Sierra 10.13.6&quot;                                                                               
##  [4] &quot;&quot;                                                                                                                       
##  [5] &quot;Matrix products: default&quot;                                                                                               
##  [6] &quot;BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib&quot;
##  [7] &quot;LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib&quot;                                    
##  [8] &quot;&quot;                                                                                                                       
##  [9] &quot;locale:&quot;                                                                                                                
## [10] &quot;[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8&quot;                                                      
## [11] &quot;&quot;                                                                                                                       
## [12] &quot;attached base packages:&quot;                                                                                                
## [13] &quot;[1] stats     graphics  grDevices utils     datasets  methods   base     &quot;                                              
## [14] &quot;&quot;                                                                                                                       
## [15] &quot;other attached packages:&quot;                                                                                               
## [16] &quot;[1] SeuratObject_4.0.1 Seurat_4.0.2      &quot;                                                                              
## [17] &quot;&quot;                                                                                                                       
## [18] &quot;loaded via a namespace (and not attached):&quot;                                                                             
## [19] &quot;  [1] Rtsne_0.15            colorspace_2.0-1      deldir_0.2-10         ellipsis_0.3.2       &quot;                          
## [20] &quot;  [5] ggridges_0.5.3        rstudioapi_0.13       spatstat.data_2.1-0   leiden_0.3.8         &quot;                          
## [21] &quot;  [9] listenv_0.8.0         ggrepel_0.9.1         fansi_0.5.0           codetools_0.2-18     &quot;                          
## [22] &quot; [13] splines_4.0.2         knitr_1.33            polyclip_1.10-0       jsonlite_1.7.2       &quot;                          
## [23] &quot; [17] packrat_0.6.0         ica_1.0-2             cluster_2.1.2         png_0.1-7            &quot;                          
## [24] &quot; [21] uwot_0.1.10           shiny_1.6.0           sctransform_0.3.2     spatstat.sparse_2.0-0&quot;                          
## [25] &quot; [25] compiler_4.0.2        httr_1.4.2            assertthat_0.2.1      Matrix_1.3-3         &quot;                          
## [26] &quot; [29] fastmap_1.1.0         lazyeval_0.2.2        later_1.2.0           htmltools_0.5.1.1    &quot;                          
## [27] &quot; [33] tools_4.0.2           igraph_1.2.6          gtable_0.3.0          glue_1.4.2           &quot;                          
## [28] &quot; [37] RANN_2.6.1            reshape2_1.4.4        dplyr_1.0.6           Rcpp_1.0.6           &quot;                          
## [29] &quot; [41] scattermore_0.7       jquerylib_0.1.4       vctrs_0.3.8           nlme_3.1-152         &quot;                          
## [30] &quot; [45] lmtest_0.9-38         xfun_0.23             stringr_1.4.0         globals_0.14.0       &quot;                          
## [31] &quot; [49] mime_0.10             miniUI_0.1.1.1        lifecycle_1.0.0       irlba_2.3.3          &quot;                          
## [32] &quot; [53] goftest_1.2-2         future_1.21.0         MASS_7.3-54           zoo_1.8-9            &quot;                          
## [33] &quot; [57] scales_1.1.1          spatstat.core_2.1-2   promises_1.2.0.1      spatstat.utils_2.1-0 &quot;                          
## [34] &quot; [61] parallel_4.0.2        RColorBrewer_1.1-2    yaml_2.2.1            reticulate_1.20      &quot;                          
## [35] &quot; [65] pbapply_1.4-3         gridExtra_2.3         ggplot2_3.3.3         sass_0.4.0           &quot;                          
## [36] &quot; [69] rpart_4.1-15          stringi_1.6.2         rlang_0.4.11          pkgconfig_2.0.3      &quot;                          
## [37] &quot; [73] matrixStats_0.59.0    evaluate_0.14         lattice_0.20-44       ROCR_1.0-11          &quot;                          
## [38] &quot; [77] purrr_0.3.4           tensor_1.5            patchwork_1.1.1       htmlwidgets_1.5.3    &quot;                          
## [39] &quot; [81] cowplot_1.1.1         tidyselect_1.1.1      parallelly_1.25.0     RcppAnnoy_0.0.18     &quot;                          
## [40] &quot; [85] plyr_1.8.6            magrittr_2.0.1        bookdown_0.22         R6_2.5.0             &quot;                          
## [41] &quot; [89] generics_0.1.0        DBI_1.1.1             mgcv_1.8-35           pillar_1.6.1         &quot;                          
## [42] &quot; [93] fitdistrplus_1.1-5    survival_3.2-11       abind_1.4-5           tibble_3.1.2         &quot;                          
## [43] &quot; [97] future.apply_1.7.0    crayon_1.4.1          KernSmooth_2.23-20    utf8_1.2.1           &quot;                          
## [44] &quot;[101] spatstat.geom_2.1-0   plotly_4.9.3          rmarkdown_2.8         grid_4.0.2           &quot;                          
## [45] &quot;[105] data.table_1.14.0     digest_0.6.27         xtable_1.8-4          tidyr_1.1.3          &quot;                          
## [46] &quot;[109] httpuv_1.6.1          munsell_0.5.0         viridisLite_0.4.0     bslib_0.2.5.1        &quot;</code></pre>
</div>
</div>
<div id="load-datasets" class="section level3">
<h3><span class="header-section-number">1.0.2</span> LOAD DATASETS</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co">###Load datasets from local (MTX files produced by Cell Ranger) and create Seurat objects</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">SeuratObjectsUnfiltered &lt;-<span class="kw">list</span>()</a>
<a class="sourceLine" id="cb12-6" data-line-number="6"></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="cf">for</span> (datasetNumber <span class="cf">in</span> <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(DatasetIds))) {</a>
<a class="sourceLine" id="cb12-8" data-line-number="8">  dataset &lt;-<span class="st"> </span>DatasetIds[[<span class="kw">as.character</span>(datasetNumber)]]</a>
<a class="sourceLine" id="cb12-9" data-line-number="9">  DatasetPath &lt;-<span class="st"> </span>PathToDatasets[[dataset]]</a>
<a class="sourceLine" id="cb12-10" data-line-number="10">  DatasetPath&lt;-<span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), DatasetPath)</a>
<a class="sourceLine" id="cb12-11" data-line-number="11">  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Loading:&quot;</span>, datasetNumber, dataset, DatasetPath, <span class="dt">sep =</span> <span class="st">&quot; &quot;</span>))</a>
<a class="sourceLine" id="cb12-12" data-line-number="12">  </a>
<a class="sourceLine" id="cb12-13" data-line-number="13">  <span class="co">### Create sparse matrix</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14">  expression_matrix.mat &lt;-<span class="st"> </span><span class="kw">Read10X</span>(<span class="dt">data.dir =</span> DatasetPath, <span class="dt">strip.suffix =</span> T)</a>
<a class="sourceLine" id="cb12-15" data-line-number="15">  <span class="kw">colnames</span>(expression_matrix.mat) &lt;-<span class="st"> </span><span class="kw">paste</span>(dataset, <span class="kw">colnames</span>(expression_matrix.mat), <span class="dt">sep =</span> <span class="st">&quot;_&quot;</span>) <span class="co">### Add datasetID to cell-barcode</span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16">  <span class="kw">dim</span>(expression_matrix.mat)</a>
<a class="sourceLine" id="cb12-17" data-line-number="17">  </a>
<a class="sourceLine" id="cb12-18" data-line-number="18">  <span class="co">### Create Seurat object</span></a>
<a class="sourceLine" id="cb12-19" data-line-number="19">  seurat.object.u &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(<span class="dt">counts =</span> expression_matrix.mat,</a>
<a class="sourceLine" id="cb12-20" data-line-number="20">                                        <span class="dt">min.cells =</span> DefaultParameters<span class="op">$</span>MinCells, </a>
<a class="sourceLine" id="cb12-21" data-line-number="21">                                        <span class="dt">min.features =</span> DefaultParameters<span class="op">$</span>MinGenes, </a>
<a class="sourceLine" id="cb12-22" data-line-number="22">                                        <span class="dt">project =</span> <span class="kw">paste0</span>(PrefixOutfiles, <span class="st">&quot;_&quot;</span>, dataset)</a>
<a class="sourceLine" id="cb12-23" data-line-number="23">                                        )</a>
<a class="sourceLine" id="cb12-24" data-line-number="24">  seurat.object.u[[<span class="st">&#39;dataset.label&#39;</span>]] &lt;-<span class="st"> </span>dataset</a>
<a class="sourceLine" id="cb12-25" data-line-number="25">  SeuratObjectsUnfiltered[[datasetNumber]]  &lt;-<span class="st"> </span>seurat.object.u</a>
<a class="sourceLine" id="cb12-26" data-line-number="26">}</a></code></pre></div>
<pre><code>## [1] &quot;Loading: 1 G1003_A_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G1003_A_T/&quot;
## [1] &quot;Loading: 2 G620_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G620_T/&quot;
## [1] &quot;Loading: 3 G910_A_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G910_A_T/&quot;
## [1] &quot;Loading: 4 G945_I_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G945_I_T/&quot;
## [1] &quot;Loading: 5 G946_I_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G946_I_T/&quot;
## [1] &quot;Loading: 6 G967_A_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G967_A_T/&quot;
## [1] &quot;Loading: 7 G983_A_T /Users/jdiazmej/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G983_A_T/&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">print</span>(SeuratObjectsUnfiltered)</a></code></pre></div>
<pre><code>## [[1]]
## An object of class Seurat 
## 16895 features across 3905 samples within 1 assay 
## Active assay: RNA (16895 features, 0 variable features)
## 
## [[2]]
## An object of class Seurat 
## 16016 features across 1012 samples within 1 assay 
## Active assay: RNA (16016 features, 0 variable features)
## 
## [[3]]
## An object of class Seurat 
## 16037 features across 2897 samples within 1 assay 
## Active assay: RNA (16037 features, 0 variable features)
## 
## [[4]]
## An object of class Seurat 
## 15783 features across 1279 samples within 1 assay 
## Active assay: RNA (15783 features, 0 variable features)
## 
## [[5]]
## An object of class Seurat 
## 15193 features across 1510 samples within 1 assay 
## Active assay: RNA (15193 features, 0 variable features)
## 
## [[6]]
## An object of class Seurat 
## 15994 features across 3056 samples within 1 assay 
## Active assay: RNA (15994 features, 0 variable features)
## 
## [[7]]
## An object of class Seurat 
## 15964 features across 1624 samples within 1 assay 
## Active assay: RNA (15964 features, 0 variable features)</code></pre>
</div>
<div id="qc-datasets" class="section level3">
<h3><span class="header-section-number">1.0.3</span> QC DATASETS</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="co">###Filter cells based on number of genes, number of reads, and mitochondrial representation</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4">SeuratObjectsFiltered &lt;-<span class="kw">list</span>()</a>
<a class="sourceLine" id="cb16-5" data-line-number="5"></a>
<a class="sourceLine" id="cb16-6" data-line-number="6"><span class="cf">for</span> (datasetNumber <span class="cf">in</span> <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(DatasetIds))) {</a>
<a class="sourceLine" id="cb16-7" data-line-number="7">  dataset &lt;-<span class="st"> </span>DatasetIds[[<span class="kw">as.character</span>(datasetNumber)]]</a>
<a class="sourceLine" id="cb16-8" data-line-number="8">  DatasetPath &lt;-<span class="st"> </span>PathToDatasets[[dataset]]</a>
<a class="sourceLine" id="cb16-9" data-line-number="9">  seurat.object.u &lt;-<span class="st"> </span>SeuratObjectsUnfiltered[[datasetNumber]]</a>
<a class="sourceLine" id="cb16-10" data-line-number="10">  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Filtering:&quot;</span>, datasetNumber, dataset, DatasetPath, <span class="dt">sep =</span> <span class="st">&quot; &quot;</span>))</a>
<a class="sourceLine" id="cb16-11" data-line-number="11">  </a>
<a class="sourceLine" id="cb16-12" data-line-number="12">  mitoRegExpressions &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">c</span>(<span class="st">&quot;^MT-&quot;</span>), <span class="dt">collapse =</span> <span class="st">&quot;|&quot;</span>)</a>
<a class="sourceLine" id="cb16-13" data-line-number="13">  mito.features &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="dt">pattern =</span> mitoRegExpressions, </a>
<a class="sourceLine" id="cb16-14" data-line-number="14">                        <span class="dt">ignore.case =</span> T, <span class="dt">x =</span> <span class="kw">rownames</span>(<span class="dt">x =</span> SeuratObjectsUnfiltered[[datasetNumber]]), <span class="dt">value =</span> T)</a>
<a class="sourceLine" id="cb16-15" data-line-number="15"></a>
<a class="sourceLine" id="cb16-16" data-line-number="16">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-17" data-line-number="17">  <span class="co">### Get mitochondrial genes</span></a>
<a class="sourceLine" id="cb16-18" data-line-number="18">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-19" data-line-number="19">  <span class="cf">if</span> (<span class="kw">length</span>(mito.features)[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {</a>
<a class="sourceLine" id="cb16-20" data-line-number="20">    mito.fraction &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(<span class="dt">x =</span> <span class="kw">GetAssayData</span>(<span class="dt">object =</span> seurat.object.u, <span class="dt">slot =</span> <span class="st">&#39;counts&#39;</span>)[mito.features, ]) <span class="op">/</span><span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(<span class="dt">x =</span> <span class="kw">GetAssayData</span>(<span class="dt">object =</span> seurat.object.u, <span class="dt">slot =</span> <span class="st">&#39;counts&#39;</span>))</a>
<a class="sourceLine" id="cb16-21" data-line-number="21">    seurat.object.u[[<span class="st">&#39;mito.fraction&#39;</span>]] &lt;-<span class="st"> </span>mito.fraction</a>
<a class="sourceLine" id="cb16-22" data-line-number="22">  }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb16-23" data-line-number="23">    mito.fraction &lt;-<span class="st"> </span><span class="dv">0</span> <span class="op">/</span><span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(<span class="dt">x =</span> <span class="kw">GetAssayData</span>(<span class="dt">object =</span> seurat.object.u, <span class="dt">slot =</span> <span class="st">&#39;counts&#39;</span>))</a>
<a class="sourceLine" id="cb16-24" data-line-number="24">    seurat.object.u[[<span class="st">&#39;mito.fraction&#39;</span>]] &lt;-<span class="st"> </span>mito.fraction</a>
<a class="sourceLine" id="cb16-25" data-line-number="25">  }</a>
<a class="sourceLine" id="cb16-26" data-line-number="26">  </a>
<a class="sourceLine" id="cb16-27" data-line-number="27">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-28" data-line-number="28">  <span class="co">### Generate violin plots</span></a>
<a class="sourceLine" id="cb16-29" data-line-number="29">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-30" data-line-number="30">  VlnPlotPdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_QC_ViolinPlots_&quot;</span>, dataset, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb16-31" data-line-number="31">  <span class="kw">pdf</span>(<span class="dt">file=</span>VlnPlotPdf, <span class="dt">width =</span> <span class="dv">10</span>, <span class="dt">height =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb16-32" data-line-number="32">  <span class="kw">print</span>(<span class="kw">VlnPlot</span>(seurat.object.u, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;mito.fraction&quot;</span>), <span class="dt">ncol =</span> <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb16-33" data-line-number="33">  <span class="kw">dev.off</span>()</a>
<a class="sourceLine" id="cb16-34" data-line-number="34">  <span class="kw">print</span>(<span class="kw">VlnPlot</span>(seurat.object.u, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;mito.fraction&quot;</span>), <span class="dt">ncol =</span> <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb16-35" data-line-number="35"></a>
<a class="sourceLine" id="cb16-36" data-line-number="36">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-37" data-line-number="37">  <span class="co">### Filter cells based gene counts, number of reads, ribosomal and mitochondrial representation</span></a>
<a class="sourceLine" id="cb16-38" data-line-number="38">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb16-39" data-line-number="39">  <span class="cf">if</span> (<span class="kw">length</span>(mito.features)[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {</a>
<a class="sourceLine" id="cb16-40" data-line-number="40">    seurat.object.f&lt;-<span class="kw">subset</span>(<span class="dt">x =</span> seurat.object.u, <span class="dt">subset =</span> </a>
<a class="sourceLine" id="cb16-41" data-line-number="41">                              nFeature_RNA  <span class="op">&gt;=</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MinGenes</a>
<a class="sourceLine" id="cb16-42" data-line-number="42">                            <span class="op">&amp;</span><span class="st"> </span>nCount_RNA    <span class="op">&lt;=</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MaxReads</a>
<a class="sourceLine" id="cb16-43" data-line-number="43">                            <span class="op">&amp;</span><span class="st"> </span>mito.fraction <span class="op">&lt;=</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MaxMitoFraction)</a>
<a class="sourceLine" id="cb16-44" data-line-number="44">    </a>
<a class="sourceLine" id="cb16-45" data-line-number="45">  }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb16-46" data-line-number="46">    seurat.object.f&lt;-<span class="kw">subset</span>(<span class="dt">x =</span> seurat.object.u, <span class="dt">subset =</span> </a>
<a class="sourceLine" id="cb16-47" data-line-number="47">                              nFeature_RNA  <span class="op">&gt;=</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MinGenes</a>
<a class="sourceLine" id="cb16-48" data-line-number="48">                            <span class="op">&amp;</span><span class="st"> </span>nCount_RNA    <span class="op">&lt;=</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MaxReads)</a>
<a class="sourceLine" id="cb16-49" data-line-number="49">  }</a>
<a class="sourceLine" id="cb16-50" data-line-number="50">  </a>
<a class="sourceLine" id="cb16-51" data-line-number="51">  SeuratObjectsFiltered[[datasetNumber]]  &lt;-<span class="st"> </span>seurat.object.f</a>
<a class="sourceLine" id="cb16-52" data-line-number="52"></a>
<a class="sourceLine" id="cb16-53" data-line-number="53">}</a></code></pre></div>
<pre><code>## [1] &quot;Filtering: 1 G1003_A_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G1003_A_T/&quot;</code></pre>
<pre><code>## [1] &quot;Filtering: 2 G620_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G620_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<pre><code>## [1] &quot;Filtering: 3 G910_A_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G910_A_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-2.png" width="672" /></p>
<pre><code>## [1] &quot;Filtering: 4 G945_I_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G945_I_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-3.png" width="672" /></p>
<pre><code>## [1] &quot;Filtering: 5 G946_I_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G946_I_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-4.png" width="672" /></p>
<pre><code>## [1] &quot;Filtering: 6 G967_A_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G967_A_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-5.png" width="672" /></p>
<pre><code>## [1] &quot;Filtering: 7 G983_A_T ~/CourseData/CAN_data/Module7/MTX_FILES/Richards_NatCancer_2021/G983_A_T/&quot;</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-9-6.png" width="672" /><img src="bookdownproj_files/figure-html/unnamed-chunk-9-7.png" width="672" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co">### Merge Seurat objects RNA assay</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb24-4" data-line-number="4">FirstSeuratObject   &lt;-<span class="st"> </span>SeuratObjectsFiltered[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb24-5" data-line-number="5">RestOfSeuratObjectsFiltered &lt;-<span class="st"> </span>SeuratObjectsFiltered[<span class="kw">c</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(DatasetIds))]</a>
<a class="sourceLine" id="cb24-6" data-line-number="6">RestOfDatasetsIds    &lt;-<span class="st"> </span><span class="kw">unlist</span>(DatasetIds[<span class="kw">c</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(DatasetIds))])</a>
<a class="sourceLine" id="cb24-7" data-line-number="7"></a>
<a class="sourceLine" id="cb24-8" data-line-number="8">seurat.object.merged &lt;-<span class="st"> </span><span class="kw">merge</span>(FirstSeuratObject, <span class="dt">y =</span> RestOfSeuratObjectsFiltered, <span class="dt">project =</span> PrefixOutfiles)</a>
<a class="sourceLine" id="cb24-9" data-line-number="9">seurat.object.merged &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat.object.merged, <span class="dt">metadata =</span> seurat.object.merged<span class="op">@</span>meta.data<span class="op">$</span>dataset.label, <span class="dt">col.name =</span> <span class="st">&quot;dataset&quot;</span>)</a>
<a class="sourceLine" id="cb24-10" data-line-number="10">seurat.object.list &lt;-<span class="st"> </span><span class="kw">SplitObject</span>(seurat.object.merged, <span class="dt">split.by =</span> <span class="st">&quot;dataset.label&quot;</span>)</a>
<a class="sourceLine" id="cb24-11" data-line-number="11"></a>
<a class="sourceLine" id="cb24-12" data-line-number="12"><span class="kw">print</span>(seurat.object.list)</a></code></pre></div>
<pre><code>## $G1003_A_T
## An object of class Seurat 
## 17974 features across 3689 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G620_T
## An object of class Seurat 
## 17974 features across 964 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G910_A_T
## An object of class Seurat 
## 17974 features across 2815 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G945_I_T
## An object of class Seurat 
## 17974 features across 662 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G946_I_T
## An object of class Seurat 
## 17974 features across 1447 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G967_A_T
## An object of class Seurat 
## 17974 features across 3019 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)
## 
## $G983_A_T
## An object of class Seurat 
## 17974 features across 1536 samples within 1 assay 
## Active assay: RNA (17974 features, 0 variable features)</code></pre>
</div>
<div id="normalize-datasets" class="section level3">
<h3><span class="header-section-number">1.0.4</span> NORMALIZE DATASETS</h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="co">### Running SCTransform</span></a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(seurat.object.list)) {</a>
<a class="sourceLine" id="cb26-5" data-line-number="5">  dataset &lt;-<span class="st"> </span><span class="kw">names</span>(seurat.object.list)[[i]]</a>
<a class="sourceLine" id="cb26-6" data-line-number="6">  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Normalizing:&quot;</span>, i, dataset, <span class="dt">sep =</span> <span class="st">&quot; &quot;</span>))</a>
<a class="sourceLine" id="cb26-7" data-line-number="7">  seurat.object.list[[i]] &lt;-<span class="st"> </span><span class="kw">SCTransform</span>(seurat.object.list[[i]], <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb26-8" data-line-number="8">}</a></code></pre></div>
<pre><code>## [1] &quot;Normalizing: 1 G1003_A_T&quot;
## [1] &quot;Normalizing: 2 G620_T&quot;
## [1] &quot;Normalizing: 3 G910_A_T&quot;
## [1] &quot;Normalizing: 4 G945_I_T&quot;
## [1] &quot;Normalizing: 5 G946_I_T&quot;
## [1] &quot;Normalizing: 6 G967_A_T&quot;
## [1] &quot;Normalizing: 7 G983_A_T&quot;</code></pre>
</div>
<div id="save-each-dataset-r_object" class="section level3">
<h3><span class="header-section-number">1.0.5</span> SAVE EACH DATASET R_OBJECT</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="co">### Saving each dataset R object</span></a>
<a class="sourceLine" id="cb28-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb28-4" data-line-number="4"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(seurat.object.list)) {</a>
<a class="sourceLine" id="cb28-5" data-line-number="5">  dataset &lt;-<span class="st"> </span><span class="kw">names</span>(seurat.object.list)[[i]]</a>
<a class="sourceLine" id="cb28-6" data-line-number="6">  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Saving:&quot;</span>, i, dataset, <span class="dt">sep =</span> <span class="st">&quot; &quot;</span>))</a>
<a class="sourceLine" id="cb28-7" data-line-number="7">  OutfileRDS&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, dataset , <span class="st">&quot;_QC_Normalization.rds&quot;</span>)</a>
<a class="sourceLine" id="cb28-8" data-line-number="8">  <span class="kw">saveRDS</span>(seurat.object.list[[i]], <span class="dt">file =</span> OutfileRDS)</a>
<a class="sourceLine" id="cb28-9" data-line-number="9">}</a></code></pre></div>
<pre><code>## [1] &quot;Saving: 1 G1003_A_T&quot;
## [1] &quot;Saving: 2 G620_T&quot;
## [1] &quot;Saving: 3 G910_A_T&quot;
## [1] &quot;Saving: 4 G945_I_T&quot;
## [1] &quot;Saving: 5 G946_I_T&quot;
## [1] &quot;Saving: 6 G967_A_T&quot;
## [1] &quot;Saving: 7 G983_A_T&quot;</code></pre>
</div>
<div id="obtain-computing-time" class="section level3">
<h3><span class="header-section-number">1.0.6</span> OBTAIN COMPUTING TIME</h3>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="co">### Obtain computing time used</span></a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4">StopWatchEnd<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a>
<a class="sourceLine" id="cb30-5" data-line-number="5"></a>
<a class="sourceLine" id="cb30-6" data-line-number="6">OutfileCPUtimes&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_QC_Normalization.txt&quot;</span>)</a>
<a class="sourceLine" id="cb30-7" data-line-number="7"></a>
<a class="sourceLine" id="cb30-8" data-line-number="8">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Step&quot;</span>, <span class="st">&quot;Time(minutes)&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb30-9" data-line-number="9"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> OutfileCPUtimes, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb30-10" data-line-number="10"></a>
<a class="sourceLine" id="cb30-11" data-line-number="11"><span class="kw">lapply</span>(<span class="kw">names</span>(StopWatchStart), <span class="cf">function</span>(stepToClock) {</a>
<a class="sourceLine" id="cb30-12" data-line-number="12">  <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;POSIXct&quot;</span>, <span class="kw">class</span>(StopWatchStart[[stepToClock]]), <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb30-13" data-line-number="13">    TimeStart &lt;-<span class="st"> </span>StopWatchStart[[stepToClock]]</a>
<a class="sourceLine" id="cb30-14" data-line-number="14">    TimeEnd   &lt;-<span class="st"> </span>StopWatchEnd[[stepToClock]]</a>
<a class="sourceLine" id="cb30-15" data-line-number="15">    TimeDiff &lt;-<span class="st"> </span><span class="kw">format</span>(<span class="kw">difftime</span>(TimeEnd, TimeStart, <span class="dt">units =</span> <span class="st">&quot;min&quot;</span>))</a>
<a class="sourceLine" id="cb30-16" data-line-number="16">    ReportTime&lt;-<span class="kw">c</span>(<span class="kw">paste</span>(stepToClock, TimeDiff, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb30-17" data-line-number="17">    <span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x=</span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot; mins&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">x =</span> ReportTime), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb30-18" data-line-number="18">  }</a>
<a class="sourceLine" id="cb30-19" data-line-number="19">})</a></code></pre></div>
<pre><code>## [[1]]
## NULL</code></pre>
</div>
<div id="finish" class="section level3">
<h3><span class="header-section-number">1.0.7</span> FINISH</h3>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="co">### Finish</span></a>
<a class="sourceLine" id="cb32-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb32-4" data-line-number="4"><span class="kw">options</span>(<span class="dt">warn =</span> oldw)</a>
<a class="sourceLine" id="cb32-5" data-line-number="5"><span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">END - Check:</span><span class="ch">\n</span><span class="st">&quot;</span>, PathForOutfiles, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">For outfiles</span><span class="ch">\n\n</span><span class="st">&quot;</span>))</a></code></pre></div>
<pre><code>## 
## END - Check:
## /Users/jdiazmej/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021
## For outfiles</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co"># quit()</span></a></code></pre></div>

<ul>
<li>Javier Diaz - <a href="mailto:javier.diazmejia@gmail.com" class="email">javier.diazmejia@gmail.com</a></li>
<li><p>Script based on: <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="uri">https://satijalab.org/seurat/articles/integration_introduction.html</a></p></li>
<li>GENERAL OVERVIEW OF THIS SCRIPT
<ol style="list-style-type: decimal">
<li>Loads each normalized dataset R object produced by script CBW_CAN_2021_Module7_Lab1_QC_Normalization.R</li>
<li>Merges and integrates datasets correcting batch effects</li>
<li>Saves R object with integrated datasets</li>
</ol></li>
</ul>
<hr />
</div>
<div id="define-environment-inputs-and-outdir-1" class="section level3">
<h3><span class="header-section-number">1.0.8</span> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1">oldw &lt;-<span class="st"> </span><span class="kw">getOption</span>(<span class="st">&quot;warn&quot;</span>)</a>
<a class="sourceLine" id="cb35-2" data-line-number="2"><span class="kw">options</span>( <span class="dt">warn =</span> <span class="dv">-1</span> )</a></code></pre></div>
<div id="required-libraries-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.1</span> Required libraries</h6>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">library</span>(Seurat) <span class="co"># (CRAN) main scRNA-seq analysis package</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="kw">library</span>(future) <span class="co"># (CRAN) to run parallel processes</span></a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="kw">library</span>(STACAS) <span class="co"># (GitHub carmonalab/STACAS) for STACAS-based anchor detection</span></a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="co">#comment</span></a></code></pre></div>
</div>
<div id="users-home-and-stopwatch-to-time-run-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.2</span> User’s home and stopwatch to time run</h6>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">UserHomeDirectory &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb37-2" data-line-number="2"><span class="kw">print</span>(UserHomeDirectory)</a></code></pre></div>
<pre><code>## [1] &quot;/Users/jdiazmej&quot;</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1">StopWatchStart &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb39-2" data-line-number="2">StopWatchEnd   &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb39-3" data-line-number="3"></a>
<a class="sourceLine" id="cb39-4" data-line-number="4">StopWatchStart<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a></code></pre></div>
</div>
<div id="define-inputs-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.3</span> Define inputs</h6>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="co">### Run specific inputs and parameters</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2"></a>
<a class="sourceLine" id="cb40-3" data-line-number="3">AnchorFinder        &lt;-<span class="st"> &quot;STACAS&quot;</span> <span class="co">## Either &#39;STACAS&#39; or &#39;SEURAT&#39;</span></a>
<a class="sourceLine" id="cb40-4" data-line-number="4"></a>
<a class="sourceLine" id="cb40-5" data-line-number="5"><span class="co">### /path_to/*rds infiles (R objects) from CBW_CAN_2021_Module7_Lab1__QC_Normalization.R</span></a>
<a class="sourceLine" id="cb40-6" data-line-number="6"></a>
<a class="sourceLine" id="cb40-7" data-line-number="7"><span class="co">### Not including G945_I_T due to memory constrains and because it has high mito.fraction</span></a>
<a class="sourceLine" id="cb40-8" data-line-number="8"></a>
<a class="sourceLine" id="cb40-9" data-line-number="9">PathToDatasets &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb40-10" data-line-number="10">  <span class="st">&quot;G1003_A_T&quot;</span> =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G1003_A_T_QC_Normalization.rds&quot;</span>,</a>
<a class="sourceLine" id="cb40-11" data-line-number="11">  <span class="co"># &quot;G945_I_T&quot;  = &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G945_I_T_QC_Normalization.rds&quot;,</span></a>
<a class="sourceLine" id="cb40-12" data-line-number="12">  <span class="st">&quot;G910_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G910_A_T_QC_Normalization.rds&quot;</span>,</a>
<a class="sourceLine" id="cb40-13" data-line-number="13">  <span class="st">&quot;G983_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G983_A_T_QC_Normalization.rds&quot;</span>,</a>
<a class="sourceLine" id="cb40-14" data-line-number="14">  <span class="st">&quot;G620_T&quot;</span>    =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G620_T_QC_Normalization.rds&quot;</span>,</a>
<a class="sourceLine" id="cb40-15" data-line-number="15">  <span class="st">&quot;G946_I_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G946_I_T_QC_Normalization.rds&quot;</span>,</a>
<a class="sourceLine" id="cb40-16" data-line-number="16">  <span class="st">&quot;G967_A_T&quot;</span>  =<span class="st"> &quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/Richards_NatCancer_2021_G967_A_T_QC_Normalization.rds&quot;</span></a>
<a class="sourceLine" id="cb40-17" data-line-number="17">)</a></code></pre></div>
</div>
<div id="define-outputs-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.4</span> Define outputs</h6>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="co">### Outputs</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>) <span class="co">## /path_to/out_directory</span></a>
<a class="sourceLine" id="cb41-3" data-line-number="3">PrefixOutfiles      &lt;-<span class="st"> &quot;Richards_NatCancer_2021&quot;</span>  <span class="co">## Prefix for outfiles</span></a>
<a class="sourceLine" id="cb41-4" data-line-number="4">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathForOutfiles)</a>
<a class="sourceLine" id="cb41-5" data-line-number="5"><span class="kw">dir.create</span>(<span class="dt">path =</span> PathForOutfiles, <span class="dt">recursive =</span> T, <span class="dt">showWarnings =</span> F)</a></code></pre></div>
</div>
<div id="define-default-parameters-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.5</span> Define default parameters</h6>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co">### Default parameters. Either suggested by Seurat developers, or tailored empirically.</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2"></a>
<a class="sourceLine" id="cb42-3" data-line-number="3"><span class="co">### </span><span class="al">NOTE</span><span class="co">: when datasets being integrated have fewer anchors than needed for the integration,</span></a>
<a class="sourceLine" id="cb42-4" data-line-number="4"><span class="co">### step IntegrateData() will produce an error:</span></a>
<a class="sourceLine" id="cb42-5" data-line-number="5"><span class="co">### `Error in nn2(data = c(15.4534704633918, 15.4534704633918, 15.4534704633918,  : </span></a>
<a class="sourceLine" id="cb42-6" data-line-number="6"><span class="co">### Cannot find more nearest neighbours than there are points`</span></a>
<a class="sourceLine" id="cb42-7" data-line-number="7"><span class="co">### To fix this using Seurat&#39;s anchor finder in step FindIntegrationAnchors() lower the k.filter value (e.g. 150, default = 200)</span></a>
<a class="sourceLine" id="cb42-8" data-line-number="8"><span class="co">### To fix this using STACAS anchor finder, in step FilterAnchors.STACAS() increase the dist.pct and dist.thr values (e.g. 0.9, default = 0.8)</span></a>
<a class="sourceLine" id="cb42-9" data-line-number="9"><span class="co">### More info: https://github.com/satijalab/seurat/issues/2056</span></a>
<a class="sourceLine" id="cb42-10" data-line-number="10">DefaultParameters &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb42-11" data-line-number="11">  </a>
<a class="sourceLine" id="cb42-12" data-line-number="12">  <span class="co">### Parameters for dataset integration</span></a>
<a class="sourceLine" id="cb42-13" data-line-number="13">  <span class="dt">IntegrationNFeatures =</span> <span class="dv">3000</span>,</a>
<a class="sourceLine" id="cb42-14" data-line-number="14">  <span class="dt">StacasVarGenesIntegratedN =</span> <span class="dv">500</span>,</a>
<a class="sourceLine" id="cb42-15" data-line-number="15">  <span class="dt">PcaDimsUse =</span> <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>),</a>
<a class="sourceLine" id="cb42-16" data-line-number="16">  <span class="dt">SeuratFindAnchorsKFilter =</span> <span class="dv">200</span>, <span class="co">#default 200</span></a>
<a class="sourceLine" id="cb42-17" data-line-number="17">  <span class="dt">StacasDistPct =</span> <span class="fl">0.9</span>, <span class="co">#default 0.8</span></a>
<a class="sourceLine" id="cb42-18" data-line-number="18">  <span class="dt">StacasDistThr =</span> <span class="fl">0.9</span>, <span class="co">#default 0.8</span></a>
<a class="sourceLine" id="cb42-19" data-line-number="19">  </a>
<a class="sourceLine" id="cb42-20" data-line-number="20">  <span class="co">### ReferenceDatasets = either a &lt;comma&gt; delimited list of dataset ID(s) to be used as</span></a>
<a class="sourceLine" id="cb42-21" data-line-number="21">  <span class="co">### references for anchors or &#39;NA&#39; to compare all-vs-all</span></a>
<a class="sourceLine" id="cb42-22" data-line-number="22">  <span class="dt">ReferenceDatasets =</span> <span class="st">&quot;NA&quot;</span></a>
<a class="sourceLine" id="cb42-23" data-line-number="23">)</a></code></pre></div>
</div>
<div id="report-r-sessioninfo-1" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.6</span> Report R sessionInfo</h6>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1">OutfileRSessionInfo&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_SingleCell_2_Integration_RSessionInfo.txt&quot;</span>)</a>
<a class="sourceLine" id="cb43-2" data-line-number="2"><span class="kw">writeLines</span>(<span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>()), OutfileRSessionInfo)</a>
<a class="sourceLine" id="cb43-3" data-line-number="3"><span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>())</a></code></pre></div>
<pre><code>##  [1] &quot;R version 4.0.2 (2020-06-22)&quot;                                                                                           
##  [2] &quot;Platform: x86_64-apple-darwin17.0 (64-bit)&quot;                                                                             
##  [3] &quot;Running under: macOS High Sierra 10.13.6&quot;                                                                               
##  [4] &quot;&quot;                                                                                                                       
##  [5] &quot;Matrix products: default&quot;                                                                                               
##  [6] &quot;BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib&quot;
##  [7] &quot;LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib&quot;                                    
##  [8] &quot;&quot;                                                                                                                       
##  [9] &quot;locale:&quot;                                                                                                                
## [10] &quot;[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8&quot;                                                      
## [11] &quot;&quot;                                                                                                                       
## [12] &quot;attached base packages:&quot;                                                                                                
## [13] &quot;[1] stats     graphics  grDevices utils     datasets  methods   base     &quot;                                              
## [14] &quot;&quot;                                                                                                                       
## [15] &quot;other attached packages:&quot;                                                                                               
## [16] &quot;[1] STACAS_1.1.0       future_1.21.0      SeuratObject_4.0.1 Seurat_4.0.2      &quot;                                        
## [17] &quot;&quot;                                                                                                                       
## [18] &quot;loaded via a namespace (and not attached):&quot;                                                                             
## [19] &quot;  [1] Rtsne_0.15            colorspace_2.0-1      deldir_0.2-10         ellipsis_0.3.2       &quot;                          
## [20] &quot;  [5] ggridges_0.5.3        rstudioapi_0.13       spatstat.data_2.1-0   farver_2.1.0         &quot;                          
## [21] &quot;  [9] leiden_0.3.8          listenv_0.8.0         ggrepel_0.9.1         fansi_0.5.0          &quot;                          
## [22] &quot; [13] codetools_0.2-18      splines_4.0.2         knitr_1.33            polyclip_1.10-0      &quot;                          
## [23] &quot; [17] jsonlite_1.7.2        packrat_0.6.0         ica_1.0-2             cluster_2.1.2        &quot;                          
## [24] &quot; [21] png_0.1-7             uwot_0.1.10           shiny_1.6.0           sctransform_0.3.2    &quot;                          
## [25] &quot; [25] spatstat.sparse_2.0-0 compiler_4.0.2        httr_1.4.2            assertthat_0.2.1     &quot;                          
## [26] &quot; [29] Matrix_1.3-3          fastmap_1.1.0         lazyeval_0.2.2        later_1.2.0          &quot;                          
## [27] &quot; [33] htmltools_0.5.1.1     tools_4.0.2           igraph_1.2.6          gtable_0.3.0         &quot;                          
## [28] &quot; [37] glue_1.4.2            RANN_2.6.1            reshape2_1.4.4        dplyr_1.0.6          &quot;                          
## [29] &quot; [41] Rcpp_1.0.6            scattermore_0.7       jquerylib_0.1.4       vctrs_0.3.8          &quot;                          
## [30] &quot; [45] nlme_3.1-152          lmtest_0.9-38         xfun_0.23             stringr_1.4.0        &quot;                          
## [31] &quot; [49] globals_0.14.0        mime_0.10             miniUI_0.1.1.1        lifecycle_1.0.0      &quot;                          
## [32] &quot; [53] irlba_2.3.3           goftest_1.2-2         MASS_7.3-54           zoo_1.8-9            &quot;                          
## [33] &quot; [57] scales_1.1.1          spatstat.core_2.1-2   promises_1.2.0.1      spatstat.utils_2.1-0 &quot;                          
## [34] &quot; [61] parallel_4.0.2        RColorBrewer_1.1-2    yaml_2.2.1            reticulate_1.20      &quot;                          
## [35] &quot; [65] pbapply_1.4-3         gridExtra_2.3         ggplot2_3.3.3         sass_0.4.0           &quot;                          
## [36] &quot; [69] rpart_4.1-15          stringi_1.6.2         highr_0.9             rlang_0.4.11         &quot;                          
## [37] &quot; [73] pkgconfig_2.0.3       matrixStats_0.59.0    evaluate_0.14         lattice_0.20-44      &quot;                          
## [38] &quot; [77] ROCR_1.0-11           purrr_0.3.4           tensor_1.5            labeling_0.4.2       &quot;                          
## [39] &quot; [81] patchwork_1.1.1       htmlwidgets_1.5.3     cowplot_1.1.1         tidyselect_1.1.1     &quot;                          
## [40] &quot; [85] parallelly_1.25.0     RcppAnnoy_0.0.18      plyr_1.8.6            magrittr_2.0.1       &quot;                          
## [41] &quot; [89] bookdown_0.22         R6_2.5.0              generics_0.1.0        DBI_1.1.1            &quot;                          
## [42] &quot; [93] withr_2.4.2           mgcv_1.8-35           pillar_1.6.1          fitdistrplus_1.1-5   &quot;                          
## [43] &quot; [97] survival_3.2-11       abind_1.4-5           tibble_3.1.2          future.apply_1.7.0   &quot;                          
## [44] &quot;[101] crayon_1.4.1          KernSmooth_2.23-20    utf8_1.2.1            spatstat.geom_2.1-0  &quot;                          
## [45] &quot;[105] plotly_4.9.3          rmarkdown_2.8         grid_4.0.2            data.table_1.14.0    &quot;                          
## [46] &quot;[109] digest_0.6.27         xtable_1.8-4          tidyr_1.1.3           httpuv_1.6.1         &quot;                          
## [47] &quot;[113] munsell_0.5.0         viridisLite_0.4.0     bslib_0.2.5.1        &quot;</code></pre>
</div>
<div id="define-number-of-cores-and-ram-for-parallelization" class="section level6">
<h6><span class="header-section-number">1.0.8.0.0.7</span> Define number of cores and RAM for parallelization</h6>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">NumbCores &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2">MaxGlobalVariables &lt;-<span class="st"> </span><span class="dv">10000</span></a>
<a class="sourceLine" id="cb45-3" data-line-number="3"></a>
<a class="sourceLine" id="cb45-4" data-line-number="4"><span class="co">### Using plan(strategy = &quot;multisession&quot;, workers = NumbCoresToUse)</span></a>
<a class="sourceLine" id="cb45-5" data-line-number="5"><span class="co">### failed in the CBW AWS instance, with error:</span></a>
<a class="sourceLine" id="cb45-6" data-line-number="6"><span class="co">###   Failed to retrieve the result of MulticoreFuture (future_lapply-1) from the forked worker (on localhost; PID 194469).</span></a>
<a class="sourceLine" id="cb45-7" data-line-number="7"><span class="co">###   Post-mortem diagnostic: No process exists with this PID, i.e. the forked localhost worker is no longer alive.</span></a>
<a class="sourceLine" id="cb45-8" data-line-number="8"><span class="co">### This seems to be related to not having enough RAM in the AWS instance allocated for the course</span></a>
<a class="sourceLine" id="cb45-9" data-line-number="9"><span class="co">### To fix this for the CBW CAN, we are:</span></a>
<a class="sourceLine" id="cb45-10" data-line-number="10"><span class="co">###   a) removing from the analysis a dataset (G945_I_T) with high mito.fraction</span></a>
<a class="sourceLine" id="cb45-11" data-line-number="11"><span class="co">###   b) Using `plan(strategy = &quot;sequential&quot;)` in script CBW_CAN_SingleCell_2_Integration.R (i.e. no parallelization)</span></a>
<a class="sourceLine" id="cb45-12" data-line-number="12"><span class="co">### </span></a>
<a class="sourceLine" id="cb45-13" data-line-number="13"><span class="co">### Notes: `options(future.globals.maxSize = MaxGlobalVariables)` sets the maximum allowed size in bytes.</span></a>
<a class="sourceLine" id="cb45-14" data-line-number="14"><span class="co">###         To set it to 10GB, you would run options(future.globals.maxSize = 10000 * 1024^2).</span></a>
<a class="sourceLine" id="cb45-15" data-line-number="15"><span class="co">###         Increasing the value of future.globals.maxSize will increase your RAM usage so set this number mindfully.</span></a>
<a class="sourceLine" id="cb45-16" data-line-number="16"><span class="co">###         https://satijalab.org/seurat/archive/v3.1/future_vignette.html</span></a>
<a class="sourceLine" id="cb45-17" data-line-number="17"><span class="co">###         https://github.com/satijalab/seurat/issues/3249</span></a>
<a class="sourceLine" id="cb45-18" data-line-number="18"></a>
<a class="sourceLine" id="cb45-19" data-line-number="19"><span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^MAX$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb45-20" data-line-number="20">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">availableCores</span>()[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb45-21" data-line-number="21">}<span class="cf">else</span> <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^[0-9]+$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb45-22" data-line-number="22">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(NumbCores)</a>
<a class="sourceLine" id="cb45-23" data-line-number="23">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb45-24" data-line-number="24">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected format for NumbCores: &quot;</span>, NumbCores, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb45-25" data-line-number="25">}</a>
<a class="sourceLine" id="cb45-26" data-line-number="26"></a>
<a class="sourceLine" id="cb45-27" data-line-number="27"><span class="cf">if</span> (NumbCoresToUse <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb45-28" data-line-number="28">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb45-29" data-line-number="29">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;sequential&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; core ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb45-30" data-line-number="30">}<span class="cf">else</span> <span class="cf">if</span> (NumbCoresToUse <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb45-31" data-line-number="31">  <span class="co"># plan(strategy = &quot;multicore&quot;, workers = NumbCoresToUse)</span></a>
<a class="sourceLine" id="cb45-32" data-line-number="32">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;multisession&quot;</span>, <span class="dt">workers =</span> NumbCoresToUse)</a>
<a class="sourceLine" id="cb45-33" data-line-number="33">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;multicore&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; cores ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb45-34" data-line-number="34">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb45-35" data-line-number="35">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected NumbCores = &quot;</span>, NumbCoresToUse))</a>
<a class="sourceLine" id="cb45-36" data-line-number="36">}</a></code></pre></div>
<pre><code>## 
## *** Running: in &#39;sequential&#39; mode with 1 core ***</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1"><span class="kw">options</span>(<span class="dt">future.globals.maxSize =</span> MaxGlobalVariables <span class="op">*</span><span class="st"> </span><span class="dv">1024</span><span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
</div>
</div>
<div id="load-r-objects" class="section level3">
<h3><span class="header-section-number">1.0.9</span> LOAD R OBJECTS</h3>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb48-2" data-line-number="2"><span class="co">### Load R Objects</span></a>
<a class="sourceLine" id="cb48-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb48-4" data-line-number="4">InputsTable &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">matrix</span>(<span class="kw">unlist</span>(PathToDatasets), <span class="dt">nrow=</span><span class="kw">length</span>(PathToDatasets), <span class="dt">byrow=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb48-5" data-line-number="5"><span class="kw">rownames</span>(InputsTable) &lt;-<span class="st"> </span><span class="kw">names</span>(PathToDatasets)</a>
<a class="sourceLine" id="cb48-6" data-line-number="6"><span class="kw">colnames</span>(InputsTable)&lt;-<span class="kw">c</span>(<span class="st">&quot;PathToRObject&quot;</span>)</a>
<a class="sourceLine" id="cb48-7" data-line-number="7">NumberOfDatasets &lt;-<span class="st"> </span><span class="kw">nrow</span>(InputsTable)</a>
<a class="sourceLine" id="cb48-8" data-line-number="8"></a>
<a class="sourceLine" id="cb48-9" data-line-number="9">seurat.object.list &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb48-10" data-line-number="10"><span class="cf">for</span> (dataset <span class="cf">in</span> <span class="kw">rownames</span>(InputsTable)) {</a>
<a class="sourceLine" id="cb48-11" data-line-number="11">  <span class="kw">print</span>(dataset)</a>
<a class="sourceLine" id="cb48-12" data-line-number="12">  DatasetIndexInInputsTable &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="dt">x =</span> <span class="kw">rownames</span>(InputsTable) <span class="op">==</span><span class="st"> </span>dataset)</a>
<a class="sourceLine" id="cb48-13" data-line-number="13">  InputRobject &lt;-<span class="st"> </span>InputsTable[dataset,<span class="st">&quot;PathToRObject&quot;</span>]</a>
<a class="sourceLine" id="cb48-14" data-line-number="14">  seurat.object.list[[DatasetIndexInInputsTable]] &lt;-<span class="st"> </span><span class="kw">readRDS</span>(InputRobject)</a>
<a class="sourceLine" id="cb48-15" data-line-number="15">}</a></code></pre></div>
<pre><code>## [1] &quot;G1003_A_T&quot;
## [1] &quot;G910_A_T&quot;
## [1] &quot;G983_A_T&quot;
## [1] &quot;G620_T&quot;
## [1] &quot;G946_I_T&quot;
## [1] &quot;G967_A_T&quot;</code></pre>
</div>
<div id="integrate-datasets" class="section level3">
<h3><span class="header-section-number">1.0.10</span> INTEGRATE DATASETS</h3>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb50-2" data-line-number="2"><span class="co">### Get reference datasets</span></a>
<a class="sourceLine" id="cb50-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb50-4" data-line-number="4"><span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^NA$&quot;</span>, DefaultParameters<span class="op">$</span>ReferenceDatasets , <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb50-5" data-line-number="5">  ReferenceDatasets.indices &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(InputsTable))</a>
<a class="sourceLine" id="cb50-6" data-line-number="6">  <span class="kw">print</span>(<span class="st">&quot;Will run all pairwise dataset comparisons&quot;</span>)</a>
<a class="sourceLine" id="cb50-7" data-line-number="7">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb50-8" data-line-number="8">  ReferenceDatasets.list &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">strsplit</span>(DefaultParameters<span class="op">$</span>ReferenceDatasets, <span class="st">&quot;,&quot;</span>))</a>
<a class="sourceLine" id="cb50-9" data-line-number="9">  NumberOfFoundReferenceDatasetIDs &lt;-<span class="st"> </span><span class="kw">sum</span>(ReferenceDatasets.list <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(InputsTable) <span class="op">==</span><span class="st"> </span>T)</a>
<a class="sourceLine" id="cb50-10" data-line-number="10">  <span class="cf">if</span> (NumberOfFoundReferenceDatasetIDs <span class="op">==</span><span class="st"> </span><span class="kw">length</span>(ReferenceDatasets.list)) {</a>
<a class="sourceLine" id="cb50-11" data-line-number="11">    ReferenceDatasets.indices &lt;-<span class="st"> </span><span class="kw">match</span>(ReferenceDatasets.list, <span class="kw">rownames</span>(InputsTable))</a>
<a class="sourceLine" id="cb50-12" data-line-number="12">  }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb50-13" data-line-number="13">    <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Requested &quot;</span>, <span class="kw">length</span>(ReferenceDatasets.list), <span class="st">&quot; datasets as references, but found &quot;</span>, NumberOfFoundReferenceDatasetIDs))</a>
<a class="sourceLine" id="cb50-14" data-line-number="14">  }</a>
<a class="sourceLine" id="cb50-15" data-line-number="15">  <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Will use datasets: &quot;</span>, <span class="kw">paste</span>(<span class="kw">as.character</span>(ReferenceDatasets.indices), <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;,&quot;</span>)))</a>
<a class="sourceLine" id="cb50-16" data-line-number="16">}</a></code></pre></div>
<pre><code>## [1] &quot;Will run all pairwise dataset comparisons&quot;</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-2" data-line-number="2"><span class="co">### Get integration anchors</span></a>
<a class="sourceLine" id="cb52-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-4" data-line-number="4"><span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^STACAS$&quot;</span>, AnchorFinder, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb52-5" data-line-number="5">  </a>
<a class="sourceLine" id="cb52-6" data-line-number="6">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-7" data-line-number="7">  <span class="co">### Get anchors with STACAS</span></a>
<a class="sourceLine" id="cb52-8" data-line-number="8">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-9" data-line-number="9">  <span class="kw">print</span>(<span class="st">&quot;Get integration anchors with STACAS&quot;</span>)</a>
<a class="sourceLine" id="cb52-10" data-line-number="10">  seurat.object.anchors.unfiltered &lt;-<span class="st"> </span><span class="kw">FindAnchors.STACAS</span>(<span class="dt">object.list =</span> seurat.object.list, <span class="dt">dims=</span>DefaultParameters<span class="op">$</span>PcaDimsUse,</a>
<a class="sourceLine" id="cb52-11" data-line-number="11">                                                         <span class="dt">anchor.features=</span>DefaultParameters<span class="op">$</span>StacasVarGenesIntegratedN, <span class="dt">reference =</span> ReferenceDatasets.indices, <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb52-12" data-line-number="12">  seurat.object.anchors &lt;-<span class="st"> </span><span class="kw">FilterAnchors.STACAS</span>(seurat.object.anchors.unfiltered, <span class="dt">dist.thr =</span> DefaultParameters<span class="op">$</span>StacasDistPct, DefaultParameters<span class="op">$</span>StacasDistThr)</a>
<a class="sourceLine" id="cb52-13" data-line-number="13">  SampleTree &lt;-<span class="st"> </span><span class="kw">SampleTree.STACAS</span>(seurat.object.anchors)</a>
<a class="sourceLine" id="cb52-14" data-line-number="14"></a>
<a class="sourceLine" id="cb52-15" data-line-number="15">} <span class="cf">else</span> <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^Seurat$&quot;</span>, AnchorFinder , <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb52-16" data-line-number="16">  </a>
<a class="sourceLine" id="cb52-17" data-line-number="17">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-18" data-line-number="18">  <span class="co">### Get anchors with Seurat</span></a>
<a class="sourceLine" id="cb52-19" data-line-number="19">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb52-20" data-line-number="20">  <span class="kw">print</span>(<span class="st">&quot;Get integration anchors with SEURAT&quot;</span>)</a>
<a class="sourceLine" id="cb52-21" data-line-number="21">  seurat.object.integratedfeatures &lt;-<span class="st"> </span><span class="kw">SelectIntegrationFeatures</span>(<span class="dt">object.list =</span> seurat.object.list, <span class="dt">nfeatures =</span> DefaultParameters<span class="op">$</span>IntegrationNFeatures, <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb52-22" data-line-number="22">  seurat.object.list &lt;-<span class="st"> </span><span class="kw">PrepSCTIntegration</span>(<span class="dt">object.list =</span> seurat.object.list, <span class="dt">anchor.features =</span> seurat.object.integratedfeatures, <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb52-23" data-line-number="23">  seurat.object.anchors &lt;-<span class="st"> </span><span class="kw">FindIntegrationAnchors</span>(<span class="dt">object.list =</span> seurat.object.list, <span class="dt">k.filter =</span> DefaultParameters<span class="op">$</span>SeuratFindAnchorsKFilter, <span class="dt">normalization.method =</span> <span class="st">&quot;SCT&quot;</span>,</a>
<a class="sourceLine" id="cb52-24" data-line-number="24">                                                  <span class="dt">dims=</span>DefaultParameters<span class="op">$</span>PcaDimsUse, <span class="dt">anchor.features =</span> seurat.object.integratedfeatures, <span class="dt">reference =</span> ReferenceDatasets.indices,</a>
<a class="sourceLine" id="cb52-25" data-line-number="25">                                                  <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb52-26" data-line-number="26">  SampleTree &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb52-27" data-line-number="27">  </a>
<a class="sourceLine" id="cb52-28" data-line-number="28">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb52-29" data-line-number="29">  <span class="kw">stop</span>(<span class="st">&quot;ERROR: unexpected anchors function&quot;</span>)</a>
<a class="sourceLine" id="cb52-30" data-line-number="30">}</a></code></pre></div>
<pre><code>## [1] &quot;Get integration anchors with STACAS&quot;</code></pre>
<pre><code>## Preparing PCA embeddings for objects...</code></pre>
<pre><code>##  1/6 2/6 3/6 4/6 5/6 6/6</code></pre>
<pre><code>## Filter anchors using distance threshold t=0.900</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="kw">print</span>(seurat.object.anchors)</a></code></pre></div>
<pre><code>## An AnchorSet object containing 15946 anchors between 6 Seurat objects 
##  This can be used as input to IntegrateData.</code></pre>
</div>
<div id="integrate-datasets-1" class="section level3">
<h3><span class="header-section-number">1.0.11</span> INTEGRATE DATASETS</h3>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb59-2" data-line-number="2"><span class="co">### Integrate datasets</span></a>
<a class="sourceLine" id="cb59-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb59-4" data-line-number="4">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">IntegrateData</span>(<span class="dt">anchorset =</span> seurat.object.anchors, <span class="dt">normalization.method =</span> <span class="st">&quot;SCT&quot;</span>, <span class="dt">sample.tree =</span> SampleTree, <span class="dt">preserve.order =</span> T, <span class="dt">verbose =</span> T)</a></code></pre></div>
<pre><code>## Merging dataset 4 into 1</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<pre><code>## Merging dataset 2 into 5</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<pre><code>## Merging dataset 3 into 1 4</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<pre><code>## Merging dataset 5 2 into 1 4 3</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<pre><code>## Merging dataset 6 into 1 4 3 5 2</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" data-line-number="1"><span class="kw">print</span>(seurat.object.integrated)</a></code></pre></div>
<pre><code>## An object of class Seurat 
## 34523 features across 13470 samples within 3 assays 
## Active assay: integrated (500 features, 500 variable features)
##  2 other assays present: RNA, SCT</code></pre>
</div>
<div id="save-integrated-r_object" class="section level3">
<h3><span class="header-section-number">1.0.12</span> SAVE INTEGRATED R_OBJECT</h3>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb87-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb87-2" data-line-number="2"><span class="co">### Saving integrated R object</span></a>
<a class="sourceLine" id="cb87-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb87-4" data-line-number="4">OutfileRDS&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_Integration.rds&quot;</span>)</a>
<a class="sourceLine" id="cb87-5" data-line-number="5"><span class="kw">saveRDS</span>(seurat.object.integrated, <span class="dt">file =</span> OutfileRDS)</a></code></pre></div>
</div>
<div id="obtain-computing-time-1" class="section level3">
<h3><span class="header-section-number">1.0.13</span> OBTAIN COMPUTING TIME</h3>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb88-2" data-line-number="2"><span class="co">### Obtain computing time used</span></a>
<a class="sourceLine" id="cb88-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb88-4" data-line-number="4">StopWatchEnd<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a>
<a class="sourceLine" id="cb88-5" data-line-number="5"></a>
<a class="sourceLine" id="cb88-6" data-line-number="6">OutfileCPUtimes&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_Integration_CPUtimes.txt&quot;</span>)</a>
<a class="sourceLine" id="cb88-7" data-line-number="7"></a>
<a class="sourceLine" id="cb88-8" data-line-number="8"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#Number_of_cores_used&quot;</span>, NumbCoresToUse, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb88-9" data-line-number="9"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#MaxGlobalVariables&quot;</span>, MaxGlobalVariables, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb88-10" data-line-number="10"></a>
<a class="sourceLine" id="cb88-11" data-line-number="11">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Step&quot;</span>, <span class="st">&quot;Time(minutes)&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb88-12" data-line-number="12"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> OutfileCPUtimes, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb88-13" data-line-number="13"></a>
<a class="sourceLine" id="cb88-14" data-line-number="14"><span class="kw">lapply</span>(<span class="kw">names</span>(StopWatchStart), <span class="cf">function</span>(stepToClock) {</a>
<a class="sourceLine" id="cb88-15" data-line-number="15">  <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;POSIXct&quot;</span>, <span class="kw">class</span>(StopWatchStart[[stepToClock]]), <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb88-16" data-line-number="16">    TimeStart &lt;-<span class="st"> </span>StopWatchStart[[stepToClock]]</a>
<a class="sourceLine" id="cb88-17" data-line-number="17">    TimeEnd   &lt;-<span class="st"> </span>StopWatchEnd[[stepToClock]]</a>
<a class="sourceLine" id="cb88-18" data-line-number="18">    TimeDiff &lt;-<span class="st"> </span><span class="kw">format</span>(<span class="kw">difftime</span>(TimeEnd, TimeStart, <span class="dt">units =</span> <span class="st">&quot;min&quot;</span>))</a>
<a class="sourceLine" id="cb88-19" data-line-number="19">    ReportTime&lt;-<span class="kw">c</span>(<span class="kw">paste</span>(stepToClock, TimeDiff, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb88-20" data-line-number="20">    <span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x=</span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot; mins&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">x =</span> ReportTime), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb88-21" data-line-number="21">  }</a>
<a class="sourceLine" id="cb88-22" data-line-number="22">})</a></code></pre></div>
<pre><code>## [[1]]
## NULL</code></pre>
</div>
<div id="finish-1" class="section level3">
<h3><span class="header-section-number">1.0.14</span> FINISH</h3>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb90-2" data-line-number="2"><span class="co">### Finish</span></a>
<a class="sourceLine" id="cb90-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb90-4" data-line-number="4"><span class="kw">options</span>(<span class="dt">warn =</span> oldw)</a>
<a class="sourceLine" id="cb90-5" data-line-number="5"><span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">END - Check:</span><span class="ch">\n</span><span class="st">&quot;</span>, PathForOutfiles, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">For outfiles</span><span class="ch">\n\n</span><span class="st">&quot;</span>))</a></code></pre></div>
<pre><code>## 
## END - Check:
## /Users/jdiazmej/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/STACAS/WITHOUT_G945_I_T
## For outfiles</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" data-line-number="1"><span class="co"># quit()</span></a></code></pre></div>

<ul>
<li>Javier Diaz - <a href="mailto:javier.diazmejia@gmail.com" class="email">javier.diazmejia@gmail.com</a></li>
<li><p>Script based on: <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="uri">https://satijalab.org/seurat/articles/integration_introduction.html</a></p></li>
<li>GENERAL OVERVIEW OF THIS SCRIPT
<ol style="list-style-type: decimal">
<li>Loads integrated datasets R object produced by script CBW_CAN_2021_Module7_Lab2_Integration.R</li>
<li>Process integrated datasets as a whole, including:</li>
</ol>
<ul>
<li>dimension reduction</li>
<li>‘global’ cell clustering</li>
<li>UMAP plots by global cell clusters, sample, requested genes and metadata</li>
<li>Violin plots</li>
<li>Dot plots</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>Saves R object with PCA, clustering and dimension reduction</li>
</ol></li>
</ul>
<hr />
</div>
<div id="define-environment-inputs-and-outdir-2" class="section level3">
<h3><span class="header-section-number">1.0.15</span> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</h3>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb93-1" data-line-number="1">oldw &lt;-<span class="st"> </span><span class="kw">getOption</span>(<span class="st">&quot;warn&quot;</span>)</a>
<a class="sourceLine" id="cb93-2" data-line-number="2"><span class="kw">options</span>( <span class="dt">warn =</span> <span class="dv">-1</span> )</a></code></pre></div>
<div id="required-libraries-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.1</span> Required libraries</h6>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="kw">library</span>(Seurat)  <span class="co"># (CRAN) main scRNA-seq analysis package</span></a>
<a class="sourceLine" id="cb94-2" data-line-number="2"><span class="kw">library</span>(future)  <span class="co"># (CRAN) to run parallel processes</span></a>
<a class="sourceLine" id="cb94-3" data-line-number="3"><span class="kw">library</span>(ggplot2) <span class="co"># (CRAN) to generate enhanced plots</span></a>
<a class="sourceLine" id="cb94-4" data-line-number="4"><span class="kw">library</span>(DropletUtils) <span class="co"># (Bioconductor) to write out MTX format files</span></a></code></pre></div>
<pre><code>## Loading required package: SingleCellExperiment</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap,
##     parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames, dirname, do.call,
##     duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect, is.unsorted,
##     lapply, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:future&#39;:
## 
##     values</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with &#39;browseVignettes()&#39;. To cite
##     Bioconductor, see &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## 
## Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     aperm, apply, rowsum</code></pre>
<pre><code>## 
## Attaching package: &#39;SummarizedExperiment&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:SeuratObject&#39;:
## 
##     Assays</code></pre>
<pre><code>## The following object is masked from &#39;package:Seurat&#39;:
## 
##     Assays</code></pre>
</div>
<div id="users-home-and-stopwatch-to-time-run-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.2</span> User’s home and stopwatch to time run</h6>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" data-line-number="1">UserHomeDirectory &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb123-2" data-line-number="2"><span class="kw">print</span>(UserHomeDirectory)</a></code></pre></div>
<pre><code>## [1] &quot;/Users/jdiazmej&quot;</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" data-line-number="1">StopWatchStart &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb125-2" data-line-number="2">StopWatchEnd   &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb125-3" data-line-number="3"></a>
<a class="sourceLine" id="cb125-4" data-line-number="4">StopWatchStart<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a></code></pre></div>
</div>
<div id="define-inputs-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.3</span> Define inputs</h6>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb126-1" data-line-number="1"><span class="co">### Run specific inputs and parameters</span></a>
<a class="sourceLine" id="cb126-2" data-line-number="2"></a>
<a class="sourceLine" id="cb126-3" data-line-number="3">AnchorFinder        &lt;-<span class="st"> &quot;STACAS&quot;</span> <span class="co">## Either &#39;STACAS&#39; or &#39;SEURAT&#39;</span></a>
<a class="sourceLine" id="cb126-4" data-line-number="4">PathToIntegratedRds &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>, <span class="st">&quot;/Richards_NatCancer_2021_Integration.rds&quot;</span>) <span class="co">### /path_to/*rds (R object) from CBW_CAN_2021_Module7_Lab2_Integration.R</span></a>
<a class="sourceLine" id="cb126-5" data-line-number="5">InfileMetadata      &lt;-<span class="st"> &quot;~/CourseData/CAN_data/Module7/METADATA/Richards_NatCancer_2021_without_G945_I_T.metadata.tsv&quot;</span></a>
<a class="sourceLine" id="cb126-6" data-line-number="6">MetedataPropsToPlot &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cell_cluster.cell_type&quot;</span>, <span class="st">&quot;cell_type&quot;</span>)</a>
<a class="sourceLine" id="cb126-7" data-line-number="7">PathToIntegratedRds &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathToIntegratedRds)</a>
<a class="sourceLine" id="cb126-8" data-line-number="8">SelectedGenes       &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;PTPRC&quot;</span>, <span class="st">&quot;MOG&quot;</span>, <span class="st">&quot;MAG&quot;</span>, <span class="st">&quot;EGFR&quot;</span>, <span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD2&quot;</span>, <span class="st">&quot;ITGAM&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;TMEM119&quot;</span>)</a>
<a class="sourceLine" id="cb126-9" data-line-number="9">ASSAY               &lt;-<span class="st"> &quot;SCT&quot;</span></a>
<a class="sourceLine" id="cb126-10" data-line-number="10">NumberOfDimensions  &lt;-<span class="st"> </span><span class="dv">20</span></a>
<a class="sourceLine" id="cb126-11" data-line-number="11">ClustersForDotPlot2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb126-12" data-line-number="12">DimRedMethodPlots   &lt;-<span class="st"> &quot;umap&quot;</span></a>
<a class="sourceLine" id="cb126-13" data-line-number="13">SampleForInferCNV   &lt;-<span class="st"> &quot;G983_A_T&quot;</span> <span class="co">## Filtered counts and cell cluster identities from this sample will be used for CBW_CAN_SingleCell_Lab6_InferCNV.R</span></a></code></pre></div>
</div>
<div id="define-outputs-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.4</span> Define outputs</h6>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" data-line-number="1"><span class="co">### Outputs</span></a>
<a class="sourceLine" id="cb127-2" data-line-number="2">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>) <span class="co">## /path_to/out_directory</span></a>
<a class="sourceLine" id="cb127-3" data-line-number="3">PrefixOutfiles      &lt;-<span class="st"> &quot;Richards_NatCancer_2021&quot;</span>  <span class="co">## Prefix for outfiles</span></a>
<a class="sourceLine" id="cb127-4" data-line-number="4">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathForOutfiles)</a>
<a class="sourceLine" id="cb127-5" data-line-number="5"><span class="kw">dir.create</span>(<span class="dt">path =</span> PathForOutfiles, <span class="dt">recursive =</span> T, <span class="dt">showWarnings =</span> F)</a></code></pre></div>
</div>
<div id="define-default-parameters-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.5</span> Define default parameters</h6>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" data-line-number="1"><span class="co">### Default parameters. Either suggested by Seurat developers, or tailored empirically.</span></a>
<a class="sourceLine" id="cb128-2" data-line-number="2">DefaultParameters &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb128-3" data-line-number="3">  </a>
<a class="sourceLine" id="cb128-4" data-line-number="4">  <span class="co">### Parameters for QC plots</span></a>
<a class="sourceLine" id="cb128-5" data-line-number="5">  <span class="dt">CellPropertiesToQC =</span> <span class="kw">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;mito.fraction&quot;</span>),</a>
<a class="sourceLine" id="cb128-6" data-line-number="6">  </a>
<a class="sourceLine" id="cb128-7" data-line-number="7">  <span class="co">### Parameters for clustering</span></a>
<a class="sourceLine" id="cb128-8" data-line-number="8">  <span class="dt">Resolution =</span> <span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb128-9" data-line-number="9"></a>
<a class="sourceLine" id="cb128-10" data-line-number="10">  <span class="co">### Parameters for datasets comparison</span></a>
<a class="sourceLine" id="cb128-11" data-line-number="11">  <span class="dt">AssaysForAverageGETables =</span> <span class="kw">c</span>(<span class="st">&quot;RNA&quot;</span>, <span class="st">&quot;SCT&quot;</span>),</a>
<a class="sourceLine" id="cb128-12" data-line-number="12">  </a>
<a class="sourceLine" id="cb128-13" data-line-number="13">  <span class="co">### Parameters for t-SNE plots</span></a>
<a class="sourceLine" id="cb128-14" data-line-number="14">  <span class="dt">MinNumberOfCellsToReducePerplexity =</span> <span class="dv">150</span>,</a>
<a class="sourceLine" id="cb128-15" data-line-number="15">  <span class="dt">ReducedPerplexity =</span> <span class="dv">7</span></a>
<a class="sourceLine" id="cb128-16" data-line-number="16">)</a>
<a class="sourceLine" id="cb128-17" data-line-number="17"></a>
<a class="sourceLine" id="cb128-18" data-line-number="18"><span class="co">### Dimension reduction methods</span></a>
<a class="sourceLine" id="cb128-19" data-line-number="19">DimensionReductionMethods&lt;-<span class="kw">list</span>()</a>
<a class="sourceLine" id="cb128-20" data-line-number="20">DimensionReductionMethods<span class="op">$</span>umap<span class="op">$</span>name &lt;-<span class="st">&quot;UMAP&quot;</span></a>
<a class="sourceLine" id="cb128-21" data-line-number="21">DimensionReductionMethods<span class="op">$</span>tsne<span class="op">$</span>name &lt;-<span class="st">&quot;TSNE&quot;</span></a>
<a class="sourceLine" id="cb128-22" data-line-number="22">DimensionReductionMethods<span class="op">$</span>umap<span class="op">$</span>run  &lt;-<span class="kw">as.function</span>(RunUMAP)</a>
<a class="sourceLine" id="cb128-23" data-line-number="23">DimensionReductionMethods<span class="op">$</span>tsne<span class="op">$</span>run  &lt;-<span class="kw">as.function</span>(RunTSNE)</a></code></pre></div>
</div>
<div id="report-r-sessioninfo-2" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.6</span> Report R sessionInfo</h6>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" data-line-number="1">OutfileRSessionInfo&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_SingleCell_3_PCA_Clustering_DimReduction_RSessionInfo.txt&quot;</span>)</a>
<a class="sourceLine" id="cb129-2" data-line-number="2"><span class="kw">writeLines</span>(<span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>()), OutfileRSessionInfo)</a>
<a class="sourceLine" id="cb129-3" data-line-number="3"><span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>())</a></code></pre></div>
<pre><code>##  [1] &quot;R version 4.0.2 (2020-06-22)&quot;                                                                                           
##  [2] &quot;Platform: x86_64-apple-darwin17.0 (64-bit)&quot;                                                                             
##  [3] &quot;Running under: macOS High Sierra 10.13.6&quot;                                                                               
##  [4] &quot;&quot;                                                                                                                       
##  [5] &quot;Matrix products: default&quot;                                                                                               
##  [6] &quot;BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib&quot;
##  [7] &quot;LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib&quot;                                    
##  [8] &quot;&quot;                                                                                                                       
##  [9] &quot;locale:&quot;                                                                                                                
## [10] &quot;[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8&quot;                                                      
## [11] &quot;&quot;                                                                                                                       
## [12] &quot;attached base packages:&quot;                                                                                                
## [13] &quot;[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     &quot;                          
## [14] &quot;&quot;                                                                                                                       
## [15] &quot;other attached packages:&quot;                                                                                               
## [16] &quot; [1] DropletUtils_1.8.0          SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.2&quot;                               
## [17] &quot; [4] DelayedArray_0.14.1         matrixStats_0.59.0          Biobase_2.48.0             &quot;                               
## [18] &quot; [7] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2         IRanges_2.22.2             &quot;                               
## [19] &quot;[10] S4Vectors_0.26.1            BiocGenerics_0.34.0         ggplot2_3.3.3              &quot;                               
## [20] &quot;[13] STACAS_1.1.0                future_1.21.0               SeuratObject_4.0.1         &quot;                               
## [21] &quot;[16] Seurat_4.0.2               &quot;                                                                                       
## [22] &quot;&quot;                                                                                                                       
## [23] &quot;loaded via a namespace (and not attached):&quot;                                                                             
## [24] &quot;  [1] plyr_1.8.6             igraph_1.2.6           lazyeval_0.2.2         splines_4.0.2         &quot;                      
## [25] &quot;  [5] BiocParallel_1.22.0    listenv_0.8.0          scattermore_0.7        digest_0.6.27         &quot;                      
## [26] &quot;  [9] htmltools_0.5.1.1      fansi_0.5.0            magrittr_2.0.1         tensor_1.5            &quot;                      
## [27] &quot; [13] cluster_2.1.2          ROCR_1.0-11            limma_3.44.3           globals_0.14.0        &quot;                      
## [28] &quot; [17] R.utils_2.10.1         spatstat.sparse_2.0-0  colorspace_2.0-1       ggrepel_0.9.1         &quot;                      
## [29] &quot; [21] xfun_0.23              dplyr_1.0.6            crayon_1.4.1           RCurl_1.98-1.3        &quot;                      
## [30] &quot; [25] jsonlite_1.7.2         spatstat.data_2.1-0    survival_3.2-11        zoo_1.8-9             &quot;                      
## [31] &quot; [29] glue_1.4.2             polyclip_1.10-0        gtable_0.3.0           zlibbioc_1.34.0       &quot;                      
## [32] &quot; [33] XVector_0.28.0         leiden_0.3.8           Rhdf5lib_1.10.1        future.apply_1.7.0    &quot;                      
## [33] &quot; [37] HDF5Array_1.16.1       abind_1.4-5            scales_1.1.1           edgeR_3.30.3          &quot;                      
## [34] &quot; [41] DBI_1.1.1              miniUI_0.1.1.1         Rcpp_1.0.6             viridisLite_0.4.0     &quot;                      
## [35] &quot; [45] xtable_1.8-4           dqrng_0.3.0            reticulate_1.20        spatstat.core_2.1-2   &quot;                      
## [36] &quot; [49] htmlwidgets_1.5.3      httr_1.4.2             RColorBrewer_1.1-2     ellipsis_0.3.2        &quot;                      
## [37] &quot; [53] ica_1.0-2              R.methodsS3_1.8.1      pkgconfig_2.0.3        farver_2.1.0          &quot;                      
## [38] &quot; [57] sass_0.4.0             uwot_0.1.10            deldir_0.2-10          locfit_1.5-9.4        &quot;                      
## [39] &quot; [61] utf8_1.2.1             tidyselect_1.1.1       labeling_0.4.2         rlang_0.4.11          &quot;                      
## [40] &quot; [65] reshape2_1.4.4         later_1.2.0            munsell_0.5.0          tools_4.0.2           &quot;                      
## [41] &quot; [69] generics_0.1.0         ggridges_0.5.3         evaluate_0.14          stringr_1.4.0         &quot;                      
## [42] &quot; [73] fastmap_1.1.0          yaml_2.2.1             goftest_1.2-2          knitr_1.33            &quot;                      
## [43] &quot; [77] fitdistrplus_1.1-5     purrr_0.3.4            RANN_2.6.1             packrat_0.6.0         &quot;                      
## [44] &quot; [81] pbapply_1.4-3          nlme_3.1-152           mime_0.10              R.oo_1.24.0           &quot;                      
## [45] &quot; [85] compiler_4.0.2         rstudioapi_0.13        plotly_4.9.3           png_0.1-7             &quot;                      
## [46] &quot; [89] spatstat.utils_2.1-0   tibble_3.1.2           bslib_0.2.5.1          stringi_1.6.2         &quot;                      
## [47] &quot; [93] highr_0.9              lattice_0.20-44        Matrix_1.3-3           vctrs_0.3.8           &quot;                      
## [48] &quot; [97] pillar_1.6.1           lifecycle_1.0.0        spatstat.geom_2.1-0    lmtest_0.9-38         &quot;                      
## [49] &quot;[101] jquerylib_0.1.4        RcppAnnoy_0.0.18       data.table_1.14.0      cowplot_1.1.1         &quot;                      
## [50] &quot;[105] bitops_1.0-7           irlba_2.3.3            httpuv_1.6.1           patchwork_1.1.1       &quot;                      
## [51] &quot;[109] R6_2.5.0               bookdown_0.22          promises_1.2.0.1       KernSmooth_2.23-20    &quot;                      
## [52] &quot;[113] gridExtra_2.3          parallelly_1.25.0      codetools_0.2-18       MASS_7.3-54           &quot;                      
## [53] &quot;[117] assertthat_0.2.1       rhdf5_2.32.4           withr_2.4.2            sctransform_0.3.2     &quot;                      
## [54] &quot;[121] GenomeInfoDbData_1.2.3 mgcv_1.8-35            grid_4.0.2             rpart_4.1-15          &quot;                      
## [55] &quot;[125] tidyr_1.1.3            rmarkdown_2.8          Rtsne_0.15             shiny_1.6.0           &quot;</code></pre>
</div>
<div id="define-number-of-cores-and-ram-for-parallelization-1" class="section level6">
<h6><span class="header-section-number">1.0.15.0.0.7</span> Define number of cores and RAM for parallelization</h6>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" data-line-number="1">NumbCores &lt;-<span class="st"> &quot;MAX&quot;</span></a>
<a class="sourceLine" id="cb131-2" data-line-number="2">MaxGlobalVariables &lt;-<span class="st"> </span><span class="dv">10000</span></a>
<a class="sourceLine" id="cb131-3" data-line-number="3"></a>
<a class="sourceLine" id="cb131-4" data-line-number="4"><span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^MAX$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb131-5" data-line-number="5">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">availableCores</span>()[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb131-6" data-line-number="6">}<span class="cf">else</span> <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^[0-9]+$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb131-7" data-line-number="7">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(NumbCores)</a>
<a class="sourceLine" id="cb131-8" data-line-number="8">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb131-9" data-line-number="9">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected format for NumbCores: &quot;</span>, NumbCores, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb131-10" data-line-number="10">}</a>
<a class="sourceLine" id="cb131-11" data-line-number="11"></a>
<a class="sourceLine" id="cb131-12" data-line-number="12"><span class="cf">if</span> (NumbCoresToUse <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb131-13" data-line-number="13">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb131-14" data-line-number="14">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;sequential&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; core ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb131-15" data-line-number="15">}<span class="cf">else</span> <span class="cf">if</span> (NumbCoresToUse <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb131-16" data-line-number="16">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;multicore&quot;</span>, <span class="dt">workers =</span> NumbCoresToUse)</a>
<a class="sourceLine" id="cb131-17" data-line-number="17">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;multicore&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; cores ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb131-18" data-line-number="18">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb131-19" data-line-number="19">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected NumbCores = &quot;</span>, NumbCoresToUse))</a>
<a class="sourceLine" id="cb131-20" data-line-number="20">}</a></code></pre></div>
<pre><code>## 
## *** Running: in &#39;multicore&#39; mode with 4 cores ***</code></pre>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" data-line-number="1"><span class="kw">options</span>(<span class="dt">future.globals.maxSize =</span> MaxGlobalVariables <span class="op">*</span><span class="st"> </span><span class="dv">1024</span><span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
</div>
</div>
<div id="load-r-object" class="section level3">
<h3><span class="header-section-number">1.0.16</span> LOAD R OBJECT</h3>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb134-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb134-2" data-line-number="2"><span class="co">### Load R Object</span></a>
<a class="sourceLine" id="cb134-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb134-4" data-line-number="4">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">readRDS</span>(PathToIntegratedRds)</a></code></pre></div>
</div>
<div id="reduce-dimensions" class="section level3">
<h3><span class="header-section-number">1.0.17</span> REDUCE DIMENSIONS</h3>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb135-2" data-line-number="2"><span class="co">### Obtaining principal components</span></a>
<a class="sourceLine" id="cb135-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb135-4" data-line-number="4">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(seurat.object.integrated, <span class="dt">verbose =</span> F)</a>
<a class="sourceLine" id="cb135-5" data-line-number="5"></a>
<a class="sourceLine" id="cb135-6" data-line-number="6"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb135-7" data-line-number="7"><span class="co">### Generate Elbow plot</span></a>
<a class="sourceLine" id="cb135-8" data-line-number="8"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb135-9" data-line-number="9">seurat.object.integrated<span class="op">@</span>active.assay &lt;-<span class="st"> </span>ASSAY</a>
<a class="sourceLine" id="cb135-10" data-line-number="10"></a>
<a class="sourceLine" id="cb135-11" data-line-number="11">ForElbowPlot&lt;-<span class="kw">ElbowPlot</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">ndims =</span> <span class="dv">50</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>)</a>
<a class="sourceLine" id="cb135-12" data-line-number="12">MaxYAxis&lt;-<span class="kw">as.integer</span>(<span class="kw">max</span>(ForElbowPlot<span class="op">$</span>data<span class="op">$</span>stdev)<span class="op">+</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb135-13" data-line-number="13"></a>
<a class="sourceLine" id="cb135-14" data-line-number="14">OutfilePCElbowPlot &lt;-<span class="st"> </span><span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>,  PrefixOutfiles, <span class="st">&quot;_PCElbowPlot&quot;</span>, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb135-15" data-line-number="15"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePCElbowPlot, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb135-16" data-line-number="16"><span class="kw">print</span>(ForElbowPlot</a>
<a class="sourceLine" id="cb135-17" data-line-number="17">      <span class="op">+</span><span class="st"> </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks =</span>  <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="dv">50</span>, <span class="dt">by=</span><span class="dv">5</span>))</a>
<a class="sourceLine" id="cb135-18" data-line-number="18">      <span class="op">+</span><span class="st"> </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="dv">50</span>, <span class="dt">by=</span><span class="dv">5</span>), <span class="dt">linetype=</span><span class="st">&#39;dotted&#39;</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb135-19" data-line-number="19">      <span class="op">+</span><span class="st"> </span><span class="kw">scale_y_continuous</span>(<span class="dt">breaks =</span>  <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> MaxYAxis, <span class="dt">by=</span><span class="fl">0.5</span>))</a>
<a class="sourceLine" id="cb135-20" data-line-number="20">      <span class="op">+</span><span class="st"> </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> MaxYAxis, <span class="dt">by=</span><span class="fl">0.5</span>), <span class="dt">linetype=</span><span class="st">&#39;dotted&#39;</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb135-21" data-line-number="21">)</a>
<a class="sourceLine" id="cb135-22" data-line-number="22"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" data-line-number="1"><span class="kw">print</span>(ForElbowPlot</a>
<a class="sourceLine" id="cb137-2" data-line-number="2">      <span class="op">+</span><span class="st"> </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks =</span>  <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="dv">50</span>, <span class="dt">by=</span><span class="dv">5</span>))</a>
<a class="sourceLine" id="cb137-3" data-line-number="3">      <span class="op">+</span><span class="st"> </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="dv">50</span>, <span class="dt">by=</span><span class="dv">5</span>), <span class="dt">linetype=</span><span class="st">&#39;dotted&#39;</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb137-4" data-line-number="4">      <span class="op">+</span><span class="st"> </span><span class="kw">scale_y_continuous</span>(<span class="dt">breaks =</span>  <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> MaxYAxis, <span class="dt">by=</span><span class="fl">0.5</span>))</a>
<a class="sourceLine" id="cb137-5" data-line-number="5">      <span class="op">+</span><span class="st"> </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> MaxYAxis, <span class="dt">by=</span><span class="fl">0.5</span>), <span class="dt">linetype=</span><span class="st">&#39;dotted&#39;</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb137-6" data-line-number="6">)</a></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-37-1.png" width="672" /></p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb138-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb138-2" data-line-number="2"><span class="co">### Run dimension reductions using integrated data</span></a>
<a class="sourceLine" id="cb138-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb138-4" data-line-number="4"><span class="kw">writeLines</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">*** Run dimension reductions using integrated data ***</span><span class="ch">\n</span><span class="st">&quot;</span>)</a></code></pre></div>
<pre><code>## 
## *** Run dimension reductions using integrated data ***</code></pre>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb140-1" data-line-number="1">PcaDimsUse &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span>NumberOfDimensions)</a>
<a class="sourceLine" id="cb140-2" data-line-number="2"></a>
<a class="sourceLine" id="cb140-3" data-line-number="3"><span class="cf">for</span> (dim_red_method <span class="cf">in</span> <span class="kw">names</span>(DimensionReductionMethods)) {</a>
<a class="sourceLine" id="cb140-4" data-line-number="4">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb140-5" data-line-number="5">  <span class="co">### Run non-linear dimension reductions using integrated data</span></a>
<a class="sourceLine" id="cb140-6" data-line-number="6">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb140-7" data-line-number="7">  <span class="co">### NOTES:</span></a>
<a class="sourceLine" id="cb140-8" data-line-number="8">  <span class="co">### In RunTSNE: if the number of cells is too small, user may get error:</span></a>
<a class="sourceLine" id="cb140-9" data-line-number="9">  <span class="co">### `Error in Rtsne.default(X = as.matrix(x = data.use), dims = dim.embed,  : Perplexity is too large.`</span></a>
<a class="sourceLine" id="cb140-10" data-line-number="10">  <span class="co">### User can try tunning down the default RunTSNE(..., perplexity=30) to say 5 or 10</span></a>
<a class="sourceLine" id="cb140-11" data-line-number="11">  <span class="co">###</span></a>
<a class="sourceLine" id="cb140-12" data-line-number="12">  <span class="co">### Also using RunTSNE(..., check_duplicates = F) to skip cases where cells happen to have the same values after PCA reduction</span></a>
<a class="sourceLine" id="cb140-13" data-line-number="13">  </a>
<a class="sourceLine" id="cb140-14" data-line-number="14">  <span class="cf">if</span> ((<span class="st">&quot;tsne&quot;</span> <span class="op">%in%</span><span class="st"> </span>dim_red_method <span class="op">==</span><span class="st"> </span>T) <span class="op">&amp;</span><span class="st"> </span>(<span class="kw">length</span>(<span class="kw">colnames</span>(seurat.object.integrated)) <span class="op">&lt;</span><span class="st"> </span>DefaultParameters<span class="op">$</span>MinNumberOfCellsToReducePerplexity)) {</a>
<a class="sourceLine" id="cb140-15" data-line-number="15">    <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">*** Using reduced perplexity = &quot;</span>, DefaultParameters<span class="op">$</span>ReducedPerplexity, <span class="st">&quot; because found &quot;</span>,  <span class="kw">length</span>(<span class="kw">colnames</span>(seurat.object.integrated)), <span class="st">&quot; cells&quot;</span>, <span class="st">&quot; ***</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb140-16" data-line-number="16">    seurat.object.integrated &lt;-<span class="st"> </span>DimensionReductionMethods[[dim_red_method]][[<span class="st">&quot;run&quot;</span>]](<span class="dt">object =</span> seurat.object.integrated, <span class="dt">dims =</span> PcaDimsUse, <span class="dt">perplexity =</span> DefaultParameters<span class="op">$</span>ReducedPerplexity, <span class="dt">check_duplicates =</span> F)</a>
<a class="sourceLine" id="cb140-17" data-line-number="17">  }<span class="cf">else</span> <span class="cf">if</span> (<span class="st">&quot;tsne&quot;</span> <span class="op">%in%</span><span class="st"> </span>dim_red_method <span class="op">==</span><span class="st"> </span>T) {</a>
<a class="sourceLine" id="cb140-18" data-line-number="18">    seurat.object.integrated &lt;-<span class="st"> </span>DimensionReductionMethods[[dim_red_method]][[<span class="st">&quot;run&quot;</span>]](<span class="dt">object =</span> seurat.object.integrated, <span class="dt">dims =</span> PcaDimsUse, <span class="dt">check_duplicates =</span> F)</a>
<a class="sourceLine" id="cb140-19" data-line-number="19">  }<span class="cf">else</span> <span class="cf">if</span> (<span class="st">&quot;umap&quot;</span> <span class="op">%in%</span><span class="st"> </span>dim_red_method <span class="op">==</span><span class="st"> </span>T) {</a>
<a class="sourceLine" id="cb140-20" data-line-number="20">    seurat.object.integrated &lt;-<span class="st"> </span>DimensionReductionMethods[[dim_red_method]][[<span class="st">&quot;run&quot;</span>]](<span class="dt">object =</span> seurat.object.integrated, <span class="dt">dims =</span> PcaDimsUse, <span class="dt">umap.method =</span> <span class="st">&quot;uwot&quot;</span>)</a>
<a class="sourceLine" id="cb140-21" data-line-number="21">  }</a>
<a class="sourceLine" id="cb140-22" data-line-number="22">  </a>
<a class="sourceLine" id="cb140-23" data-line-number="23">  <span class="co">###################################</span></a>
<a class="sourceLine" id="cb140-24" data-line-number="24">  <span class="co">### Write out dimension reduction coordinates</span></a>
<a class="sourceLine" id="cb140-25" data-line-number="25">  <span class="co">####################################</span></a>
<a class="sourceLine" id="cb140-26" data-line-number="26">  Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[dim_red_method]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_Coordinates&quot;</span>, <span class="st">&quot;.tsv.bz2&quot;</span>), <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb140-27" data-line-number="27">  Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Barcode&quot;</span>,<span class="kw">paste</span>(<span class="kw">colnames</span>(seurat.object.integrated<span class="op">@</span>reductions[[dim_red_method]]<span class="op">@</span>cell.embeddings), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>, <span class="dt">collapse=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb140-28" data-line-number="28">  <span class="kw">write.table</span>(Headers, <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb140-29" data-line-number="29">  <span class="kw">write.table</span>(seurat.object.integrated<span class="op">@</span>reductions[[dim_red_method]]<span class="op">@</span>cell.embeddings, <span class="dt">file =</span> Outfile.con,  <span class="dt">row.names =</span> T, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb140-30" data-line-number="30">  <span class="kw">close</span>(Outfile.con)</a>
<a class="sourceLine" id="cb140-31" data-line-number="31">}</a></code></pre></div>
<pre><code>## 14:30:52 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>## 14:30:52 Read 13470 rows and found 20 numeric columns</code></pre>
<pre><code>## 14:30:52 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>## 14:30:52 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>## 0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>## [----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>## **************************************************|
## 14:30:53 Writing NN index file to temp file /var/folders/6t/btp1jy1n5kx9qxp01fqkq7yc0000gp/T//RtmpWl0868/filee8424cc310f
## 14:30:53 Searching Annoy index using 4 threads, search_k = 3000
## 14:30:55 Annoy recall = 100%
## 14:30:56 Commencing smooth kNN distance calibration using 4 threads
## 14:30:57 Initializing from normalized Laplacian + noise
## 14:30:58 Commencing optimization for 200 epochs, with 577722 positive edges
## 14:31:05 Optimization finished</code></pre>
</div>
<div id="cluster-cells" class="section level3">
<h3><span class="header-section-number">1.0.18</span> CLUSTER CELLS</h3>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb148-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb148-2" data-line-number="2"><span class="co">### Globally cluster cells using integrated data</span></a>
<a class="sourceLine" id="cb148-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb148-4" data-line-number="4"><span class="kw">writeLines</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">*** Globally cluster cells using integrated data ***</span><span class="ch">\n</span><span class="st">&quot;</span>)</a></code></pre></div>
<pre><code>## 
## *** Globally cluster cells using integrated data ***</code></pre>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb150-1" data-line-number="1"><span class="kw">options</span>(<span class="dt">scipen=</span><span class="dv">10</span>) <span class="co">## Needed to avoid an &#39;Error in file(file, &quot;rt&quot;) : cannot open the connection&#39;</span></a>
<a class="sourceLine" id="cb150-2" data-line-number="2">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">dims =</span> PcaDimsUse)</a></code></pre></div>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb153-1" data-line-number="1">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">resolution =</span> DefaultParameters<span class="op">$</span>Resolution, <span class="dt">graph.name =</span> <span class="st">&quot;integrated_snn&quot;</span>)</a></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 13470
## Number of edges: 471947
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8971
## Number of communities: 14
## Elapsed time: 2 seconds</code></pre>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb155-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-2" data-line-number="2"><span class="co">### Write out cell-cluster identities</span></a>
<a class="sourceLine" id="cb155-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-4" data-line-number="4">CellBarcodes &lt;-<span class="st"> </span><span class="kw">rownames</span>(seurat.object.integrated<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb155-5" data-line-number="5">CellClusters &lt;-<span class="st"> </span>seurat.object.integrated<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters</a>
<a class="sourceLine" id="cb155-6" data-line-number="6">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Cell_barcode&quot;</span>, <span class="kw">paste0</span>(<span class="st">&quot;seurat_cluster_r&quot;</span>, DefaultParameters<span class="op">$</span>Resolution), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb155-7" data-line-number="7">clusters_data&lt;-<span class="kw">paste</span>(CellBarcodes, CellClusters, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb155-8" data-line-number="8">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_GlobalClustering_CellClusters.tsv.bz2&quot;</span>), <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb155-9" data-line-number="9"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb155-10" data-line-number="10"><span class="kw">write.table</span>(<span class="kw">data.frame</span>(clusters_data), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb155-11" data-line-number="11"><span class="kw">close</span>(Outfile.con)</a>
<a class="sourceLine" id="cb155-12" data-line-number="12"></a>
<a class="sourceLine" id="cb155-13" data-line-number="13"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-14" data-line-number="14"><span class="co">### Write out number of cells per cluster</span></a>
<a class="sourceLine" id="cb155-15" data-line-number="15"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-16" data-line-number="16">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_GlobalClustering_NumbCellsPerCluster.tsv.bz2&quot;</span>), <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb155-17" data-line-number="17">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Cluster&quot;</span>, <span class="st">&quot;Number_of_cells&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb155-18" data-line-number="18"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb155-19" data-line-number="19"><span class="kw">write.table</span>(<span class="kw">table</span>(CellClusters), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb155-20" data-line-number="20"><span class="kw">close</span>(Outfile.con)</a>
<a class="sourceLine" id="cb155-21" data-line-number="21"></a>
<a class="sourceLine" id="cb155-22" data-line-number="22"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-23" data-line-number="23"><span class="co">### Print out the integrated matrices and cell cluster identities for CBW_CAN_SingleCell_Lab6_InferCNV.R</span></a>
<a class="sourceLine" id="cb155-24" data-line-number="24"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb155-25" data-line-number="25"><span class="kw">Idents</span>(seurat.object.integrated) &lt;-<span class="st"> &quot;dataset&quot;</span></a>
<a class="sourceLine" id="cb155-26" data-line-number="26"><span class="co">### Subset sample</span></a>
<a class="sourceLine" id="cb155-27" data-line-number="27">seurat.object.integrated.subset &lt;-<span class="st"> </span><span class="kw">subset</span>(<span class="dt">x =</span> seurat.object.integrated, <span class="dt">idents =</span> SampleForInferCNV)</a>
<a class="sourceLine" id="cb155-28" data-line-number="28"><span class="kw">print</span>(seurat.object.integrated.subset)</a></code></pre></div>
<pre><code>## An object of class Seurat 
## 34523 features across 1536 samples within 3 assays 
## Active assay: SCT (16049 features, 0 variable features)
##  2 other assays present: RNA, integrated
##  3 dimensional reductions calculated: pca, umap, tsne</code></pre>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb157-1" data-line-number="1">ASSAY &lt;-<span class="st"> &quot;SCT&quot;</span></a>
<a class="sourceLine" id="cb157-2" data-line-number="2"><span class="co">### Write out count matrices in MTX format  </span></a>
<a class="sourceLine" id="cb157-3" data-line-number="3">OutdirMTX &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>,</a>
<a class="sourceLine" id="cb157-4" data-line-number="4">                    <span class="st">&quot;/INTEGRATED_MATRIX_&quot;</span>, SampleForInferCNV, <span class="st">&quot;/&quot;</span>, ASSAY, <span class="st">&quot;/MTX&quot;</span>)</a>
<a class="sourceLine" id="cb157-5" data-line-number="5"><span class="kw">dir.create</span>(<span class="kw">file.path</span>(OutdirMTX), <span class="dt">showWarnings =</span> F, <span class="dt">recursive =</span> T)</a>
<a class="sourceLine" id="cb157-6" data-line-number="6"><span class="kw">write10xCounts</span>(<span class="dt">path =</span> OutdirMTX, <span class="dt">x =</span> seurat.object.integrated.subset<span class="op">@</span>assays[[ASSAY]]<span class="op">@</span>counts, <span class="dt">gene.type=</span><span class="st">&quot;Gene Expression&quot;</span>, <span class="dt">overwrite=</span>T, <span class="dt">type=</span><span class="st">&quot;sparse&quot;</span>, <span class="dt">version=</span><span class="st">&quot;3&quot;</span>)</a>
<a class="sourceLine" id="cb157-7" data-line-number="7"></a>
<a class="sourceLine" id="cb157-8" data-line-number="8"><span class="co">### Write out cell-cluster identities</span></a>
<a class="sourceLine" id="cb157-9" data-line-number="9">CellBarcodes &lt;-<span class="st"> </span><span class="kw">rownames</span>(seurat.object.integrated.subset<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb157-10" data-line-number="10">CellClusters &lt;-<span class="st"> </span>seurat.object.integrated.subset<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters</a>
<a class="sourceLine" id="cb157-11" data-line-number="11">clusters_data&lt;-<span class="kw">paste</span>(CellBarcodes, CellClusters, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb157-12" data-line-number="12">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_GlobalClustering_CellClusters_&quot;</span>, SampleForInferCNV, <span class="st">&quot;.tsv.bz2&quot;</span>), <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb157-13" data-line-number="13"><span class="kw">write.table</span>(<span class="kw">data.frame</span>(clusters_data), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb157-14" data-line-number="14"><span class="kw">close</span>(Outfile.con)</a></code></pre></div>
</div>
<div id="generate-umap-plots" class="section level3">
<h3><span class="header-section-number">1.0.19</span> GENERATE UMAP PLOTS</h3>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb158-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb158-2" data-line-number="2"><span class="co">### Colour dimension reduction plots by global cell-clusters</span></a>
<a class="sourceLine" id="cb158-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb158-4" data-line-number="4">plots &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(seurat.object.integrated, <span class="dt">group.by =</span> <span class="kw">c</span>(<span class="st">&quot;seurat_clusters&quot;</span>), <span class="dt">combine =</span> F, <span class="dt">reduction =</span> DimRedMethodPlots, <span class="dt">label =</span> T, <span class="dt">label.size =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb158-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> plots, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;right&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">color =</span> <span class="kw">guide_legend</span>(<span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">size =</span> <span class="dv">3</span>))))</a>
<a class="sourceLine" id="cb158-6" data-line-number="6"></a>
<a class="sourceLine" id="cb158-7" data-line-number="7">OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[DimRedMethodPlots]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_GlobalClustering_ColourByCellClusters&quot;</span>, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb158-8" data-line-number="8"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">8</span>, <span class="dt">height =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb158-9" data-line-number="9"><span class="kw">print</span>(plots)</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb160-1" data-line-number="1"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb162-1" data-line-number="1"><span class="kw">print</span>(plots)</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-40-1.png" width="672" /></p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb164-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb164-2" data-line-number="2"><span class="co">### Colour dimension reduction plots by dataset</span></a>
<a class="sourceLine" id="cb164-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb164-4" data-line-number="4">plots &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(seurat.object.integrated, <span class="dt">group.by =</span> <span class="kw">c</span>(<span class="st">&quot;dataset&quot;</span>), <span class="dt">combine =</span> F, <span class="dt">reduction =</span> DimRedMethodPlots)</a>
<a class="sourceLine" id="cb164-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> plots, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;top&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">color =</span> <span class="kw">guide_legend</span>(<span class="dt">ncol =</span> <span class="dv">6</span>, <span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">size =</span> <span class="dv">3</span>))))</a>
<a class="sourceLine" id="cb164-6" data-line-number="6"></a>
<a class="sourceLine" id="cb164-7" data-line-number="7">OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[DimRedMethodPlots]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_GlobalClustering_ColourByDataset&quot;</span>, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb164-8" data-line-number="8"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb164-9" data-line-number="9"><span class="kw">print</span>(plots)</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb166-1" data-line-number="1"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb168-1" data-line-number="1"><span class="kw">print</span>(plots)</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-41-1.png" width="672" /></p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb170-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb170-2" data-line-number="2"><span class="co">### Colour dimension reduction plots for all cells by selected genes</span></a>
<a class="sourceLine" id="cb170-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb170-4" data-line-number="4">seurat.object.integrated<span class="op">@</span>active.assay &lt;-<span class="st"> &quot;SCT&quot;</span></a>
<a class="sourceLine" id="cb170-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">features =</span> SelectedGenes, <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;lightgrey&quot;</span>, <span class="st">&quot;blue&quot;</span>),</a>
<a class="sourceLine" id="cb170-6" data-line-number="6">                     <span class="dt">reduction =</span> DimRedMethodPlots, <span class="dt">order =</span> T, <span class="dt">slot =</span> <span class="st">&quot;data&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.3</span>, <span class="dt">min.cutoff =</span> <span class="st">&quot;q0.1&quot;</span>, <span class="dt">max.cutoff =</span> <span class="st">&quot;q90&quot;</span>,</a>
<a class="sourceLine" id="cb170-7" data-line-number="7">                     <span class="dt">ncol =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb170-8" data-line-number="8"></a>
<a class="sourceLine" id="cb170-9" data-line-number="9">OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[DimRedMethodPlots]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_AllCells_ColourBySelectedGenes&quot;</span>, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb170-10" data-line-number="10"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width=</span><span class="dv">8</span>, <span class="dt">height=</span><span class="dv">8</span>)</a>
<a class="sourceLine" id="cb170-11" data-line-number="11"><span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb170-12" data-line-number="12"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb172-1" data-line-number="1"><span class="kw">print</span>(plots)</a></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-42-1.png" width="960" /></p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb173-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb173-2" data-line-number="2"><span class="co">### Colour dimension reduction plots using cell-level metadata</span></a>
<a class="sourceLine" id="cb173-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb173-4" data-line-number="4"><span class="co">### Load Metadata</span></a>
<a class="sourceLine" id="cb173-5" data-line-number="5">CellPropertiesFromMetadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">read.table</span>(InfileMetadata, <span class="dt">header =</span> T, <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">check.names =</span> F))</a>
<a class="sourceLine" id="cb173-6" data-line-number="6"><span class="kw">head</span>(CellPropertiesFromMetadata)</a></code></pre></div>
<pre><code>##                            cell_cluster cell_cluster.cell_type cell_type found_in_cell_cluster
## G1003_A_T_AAACCTGAGTATCGAA            1           01.malignant malignant                     1
## G1003_A_T_AAACCTGCAGTCCTTC            9           09.malignant malignant                     1
## G1003_A_T_AAACCTGTCAGTTTGG           11           11.malignant malignant                     1
## G1003_A_T_AAACCTGTCGGAATCT            2           02.malignant malignant                     1
## G1003_A_T_AAACCTGTCGTTTAGG            2           02.malignant malignant                     1
## G1003_A_T_AAACGGGAGAAACCGC            2           02.malignant malignant                     1
##                            dataset_id monochrome
## G1003_A_T_AAACCTGAGTATCGAA  G1003_A_T          1
## G1003_A_T_AAACCTGCAGTCCTTC  G1003_A_T          1
## G1003_A_T_AAACCTGTCAGTTTGG  G1003_A_T          1
## G1003_A_T_AAACCTGTCGGAATCT  G1003_A_T          1
## G1003_A_T_AAACCTGTCGTTTAGG  G1003_A_T          1
## G1003_A_T_AAACGGGAGAAACCGC  G1003_A_T          1</code></pre>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb175-1" data-line-number="1"><span class="co">### Add Metadata to Seurat object</span></a>
<a class="sourceLine" id="cb175-2" data-line-number="2"><span class="kw">head</span>(seurat.object.integrated<span class="op">@</span>meta.data)</a></code></pre></div>
<pre><code>##                            orig.ident nCount_RNA nFeature_RNA dataset.label mito.fraction
## G1003_A_T_AAACCTGAGTATCGAA      G1003       6271         2252     G1003_A_T    0.03364695
## G1003_A_T_AAACCTGCAGTCCTTC      G1003      12496         3440     G1003_A_T    0.04793534
## G1003_A_T_AAACCTGTCAGTTTGG      G1003      11705         3734     G1003_A_T    0.10081162
## G1003_A_T_AAACCTGTCGGAATCT      G1003       7791         2931     G1003_A_T    0.12347581
## G1003_A_T_AAACCTGTCGTTTAGG      G1003       8277         3083     G1003_A_T    0.09423704
## G1003_A_T_AAACGGGAGAAACCGC      G1003       5297         2415     G1003_A_T    0.06890693
##                              dataset nCount_SCT nFeature_SCT integrated_snn_res.0.5
## G1003_A_T_AAACCTGAGTATCGAA G1003_A_T       7200         2252                      2
## G1003_A_T_AAACCTGCAGTCCTTC G1003_A_T       8470         3270                      8
## G1003_A_T_AAACCTGTCAGTTTGG G1003_A_T       8599         3666                     11
## G1003_A_T_AAACCTGTCGGAATCT G1003_A_T       7779         2931                      1
## G1003_A_T_AAACCTGTCGTTTAGG G1003_A_T       8026         3083                      1
## G1003_A_T_AAACGGGAGAAACCGC G1003_A_T       6973         2417                      1
##                            seurat_clusters
## G1003_A_T_AAACCTGAGTATCGAA               2
## G1003_A_T_AAACCTGCAGTCCTTC               8
## G1003_A_T_AAACCTGTCAGTTTGG              11
## G1003_A_T_AAACCTGTCGGAATCT               1
## G1003_A_T_AAACCTGTCGTTTAGG               1
## G1003_A_T_AAACGGGAGAAACCGC               1</code></pre>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb177-1" data-line-number="1">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">metadata =</span> CellPropertiesFromMetadata)</a>
<a class="sourceLine" id="cb177-2" data-line-number="2"></a>
<a class="sourceLine" id="cb177-3" data-line-number="3"><span class="cf">for</span> (property <span class="cf">in</span> MetedataPropsToPlot) {</a>
<a class="sourceLine" id="cb177-4" data-line-number="4">  plots &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(seurat.object.integrated, <span class="dt">group.by =</span> property, <span class="dt">combine =</span> F, <span class="dt">reduction =</span> DimRedMethodPlots, <span class="dt">label =</span> T)</a>
<a class="sourceLine" id="cb177-5" data-line-number="5">  plots &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> plots, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;top&quot;</span>, <span class="dt">legend.text.align =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> property) <span class="op">+</span></a>
<a class="sourceLine" id="cb177-6" data-line-number="6"><span class="st">                    </span><span class="kw">guides</span>(<span class="dt">color =</span> <span class="kw">guide_legend</span>(<span class="dt">ncol =</span> <span class="dv">5</span>, <span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">size =</span> <span class="dv">3</span>))))</a>
<a class="sourceLine" id="cb177-7" data-line-number="7"></a>
<a class="sourceLine" id="cb177-8" data-line-number="8">  OutfilePdf &lt;-<span class="st"> </span><span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[DimRedMethodPlots]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_Metadata_&quot;</span>, property, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb177-9" data-line-number="9">  <span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb177-10" data-line-number="10">  <span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb177-11" data-line-number="11">  <span class="kw">dev.off</span>()</a>
<a class="sourceLine" id="cb177-12" data-line-number="12">  <span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb177-13" data-line-number="13">}</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-43-1.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-43-2.png" width="672" /></p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb182-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb182-2" data-line-number="2"><span class="co">### Colour dimension reduction plots by global cell-clusters, subsetting each dataset</span></a>
<a class="sourceLine" id="cb182-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb182-4" data-line-number="4"><span class="cf">for</span> (datasetID <span class="cf">in</span> <span class="kw">unique</span>(seurat.object.integrated<span class="op">@</span>meta.data<span class="op">$</span>dataset)) {</a>
<a class="sourceLine" id="cb182-5" data-line-number="5">  <span class="kw">Idents</span>(seurat.object.integrated) &lt;-<span class="st"> &quot;dataset&quot;</span></a>
<a class="sourceLine" id="cb182-6" data-line-number="6">  seurat.object.integrated.subset &lt;-<span class="st"> </span><span class="kw">subset</span>(<span class="dt">x =</span> seurat.object.integrated, <span class="dt">idents =</span> datasetID)</a>
<a class="sourceLine" id="cb182-7" data-line-number="7"></a>
<a class="sourceLine" id="cb182-8" data-line-number="8">  plots &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(seurat.object.integrated.subset, <span class="dt">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="dt">combine =</span> F, <span class="dt">reduction =</span> DimRedMethodPlots, <span class="dt">label =</span> T)</a>
<a class="sourceLine" id="cb182-9" data-line-number="9">  plots &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> plots, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;top&quot;</span>, <span class="dt">legend.text.align =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb182-10" data-line-number="10">                  <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="kw">paste0</span>(<span class="st">&quot;seurat_clusters - &quot;</span>, datasetID))</a>
<a class="sourceLine" id="cb182-11" data-line-number="11">                  <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">color =</span> <span class="kw">guide_legend</span>(<span class="dt">ncol =</span> <span class="dv">7</span>, <span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">size =</span> <span class="dv">3</span>))))</a>
<a class="sourceLine" id="cb182-12" data-line-number="12"></a>
<a class="sourceLine" id="cb182-13" data-line-number="13">  OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, DimensionReductionMethods[[DimRedMethodPlots]][[<span class="st">&quot;name&quot;</span>]], <span class="st">&quot;Plot_&quot;</span>, datasetID, <span class="st">&quot;_GlobalClustering_ColourByCellClusters.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb182-14" data-line-number="14">  <span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb182-15" data-line-number="15">  <span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb182-16" data-line-number="16">  <span class="kw">dev.off</span>()</a>
<a class="sourceLine" id="cb182-17" data-line-number="17">  <span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb182-18" data-line-number="18">}</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-1.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-2.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-3.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-4.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-5.png" width="672" /></p>
<pre><code>## 
## [[1]]</code></pre>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-44-6.png" width="672" /></p>
</div>
<div id="generate-dot-plots" class="section level3">
<h3><span class="header-section-number">1.0.20</span> GENERATE DOT PLOTS</h3>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb195-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb195-2" data-line-number="2"><span class="co">### Dot plots: genes (x-axis) cell-properties (y-axis), all cells</span></a>
<a class="sourceLine" id="cb195-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb195-4" data-line-number="4"><span class="kw">Idents</span>(seurat.object.integrated) &lt;-<span class="st"> &quot;seurat_clusters&quot;</span></a>
<a class="sourceLine" id="cb195-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">DotPlot</span>(seurat.object.integrated, <span class="dt">features =</span> SelectedGenes, <span class="dt">assay =</span> ASSAY, <span class="dt">cluster.idents =</span> T)</a>
<a class="sourceLine" id="cb195-6" data-line-number="6">plots &lt;-<span class="st"> </span>plots <span class="op">+</span><span class="st"> </span><span class="kw">RotatedAxis</span>()</a>
<a class="sourceLine" id="cb195-7" data-line-number="7"></a>
<a class="sourceLine" id="cb195-8" data-line-number="8">OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, <span class="st">&quot;DotPlots_AllCells_SelectedGenes.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb195-9" data-line-number="9"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb195-10" data-line-number="10"><span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb195-11" data-line-number="11"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb197-1" data-line-number="1"><span class="kw">print</span>(plots)</a></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb198-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb198-2" data-line-number="2"><span class="co">### Dot plots: genes (x-axis) cell-properties (y-axis), each dataset cells</span></a>
<a class="sourceLine" id="cb198-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb198-4" data-line-number="4"><span class="kw">Idents</span>(seurat.object.integrated) &lt;-<span class="st"> &quot;seurat_clusters&quot;</span></a>
<a class="sourceLine" id="cb198-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">DotPlot</span>(seurat.object.integrated, <span class="dt">features =</span> SelectedGenes, <span class="dt">idents =</span> ClustersForDotPlot2,</a>
<a class="sourceLine" id="cb198-6" data-line-number="6">                 <span class="dt">cols =</span> <span class="kw">c</span>(<span class="st">&quot;cornflowerblue&quot;</span>, <span class="st">&quot;darkolivegreen4&quot;</span>, <span class="st">&quot;chocolate1&quot;</span>, <span class="st">&quot;aquamarine3&quot;</span>, <span class="st">&quot;burlywood3&quot;</span>, </a>
<a class="sourceLine" id="cb198-7" data-line-number="7">                          <span class="st">&quot;azure4&quot;</span>, <span class="st">&quot;brown1&quot;</span>), <span class="dt">dot.scale =</span> <span class="dv">4</span>, <span class="dt">split.by =</span> <span class="st">&quot;dataset&quot;</span>)</a>
<a class="sourceLine" id="cb198-8" data-line-number="8">plots &lt;-<span class="st"> </span>plots <span class="op">+</span><span class="st"> </span><span class="kw">RotatedAxis</span>()</a>
<a class="sourceLine" id="cb198-9" data-line-number="9"></a>
<a class="sourceLine" id="cb198-10" data-line-number="10"></a>
<a class="sourceLine" id="cb198-11" data-line-number="11">OutfilePdf&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_&quot;</span>, <span class="st">&quot;DotPlots_EachDatasetCells&quot;</span>, <span class="st">&quot;.pdf&quot;</span>)</a>
<a class="sourceLine" id="cb198-12" data-line-number="12"><span class="kw">pdf</span>(<span class="dt">file=</span>OutfilePdf, <span class="dt">width =</span> <span class="dv">7</span>, <span class="dt">height =</span> <span class="dv">9</span>)</a>
<a class="sourceLine" id="cb198-13" data-line-number="13"><span class="kw">print</span>(plots)</a>
<a class="sourceLine" id="cb198-14" data-line-number="14"><span class="kw">dev.off</span>()</a></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb200-1" data-line-number="1"><span class="kw">print</span>(plots)</a></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-46-1.png" width="960" /></p>
</div>
<div id="get-average-gene-expression-per-cluster" class="section level3">
<h3><span class="header-section-number">1.0.21</span> GET AVERAGE GENE EXPRESSION PER CLUSTER</h3>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb201-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb201-2" data-line-number="2"><span class="co">### Get average gene expression for each global cluster</span></a>
<a class="sourceLine" id="cb201-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb201-4" data-line-number="4"><span class="kw">Idents</span>(<span class="dt">object =</span> seurat.object.integrated) &lt;-<span class="st"> &quot;seurat_clusters&quot;</span></a>
<a class="sourceLine" id="cb201-5" data-line-number="5"></a>
<a class="sourceLine" id="cb201-6" data-line-number="6">cluster.averages &lt;-<span class="st"> </span><span class="kw">AverageExpression</span>(<span class="dt">object =</span> seurat.object.integrated, <span class="dt">use.scale =</span> F, <span class="dt">use.counts =</span> F, <span class="dt">assays =</span> ASSAY)</a></code></pre></div>
<pre><code>## The following functions and any applicable methods accept the dots: CreateSeuratObject</code></pre>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" data-line-number="1">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_AverageGeneExpression_GlobalClustering.tsv.bz2&quot;</span>), <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb203-2" data-line-number="2">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;AVERAGE_GENE_EXPRESSION&quot;</span>, <span class="kw">paste</span>(<span class="st">&quot;c&quot;</span>, <span class="kw">colnames</span>(cluster.averages[[ASSAY]]), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>, <span class="dt">collapse=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb203-3" data-line-number="3"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb203-4" data-line-number="4"><span class="kw">write.table</span>(<span class="kw">data.frame</span>(cluster.averages[[ASSAY]]), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> T, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb203-5" data-line-number="5"><span class="kw">close</span>(Outfile.con)</a></code></pre></div>
</div>
<div id="save-integrated-r_object-1" class="section level3">
<h3><span class="header-section-number">1.0.22</span> SAVE INTEGRATED R_OBJECT</h3>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb204-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb204-2" data-line-number="2"><span class="co">### Saving integrated R object</span></a>
<a class="sourceLine" id="cb204-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb204-4" data-line-number="4">OutfileRDS&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_PCA_Clustering_DimReduction.rds&quot;</span>)</a>
<a class="sourceLine" id="cb204-5" data-line-number="5"><span class="kw">saveRDS</span>(seurat.object.integrated, <span class="dt">file =</span> OutfileRDS)</a></code></pre></div>
</div>
<div id="obtain-computing-time-2" class="section level3">
<h3><span class="header-section-number">1.0.23</span> OBTAIN COMPUTING TIME</h3>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb205-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb205-2" data-line-number="2"><span class="co">### Obtain computing time used</span></a>
<a class="sourceLine" id="cb205-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb205-4" data-line-number="4">StopWatchEnd<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a>
<a class="sourceLine" id="cb205-5" data-line-number="5"></a>
<a class="sourceLine" id="cb205-6" data-line-number="6">OutfileCPUtimes&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_PCA_Clustering_DimReduction_CPUtimes.txt&quot;</span>)</a>
<a class="sourceLine" id="cb205-7" data-line-number="7"></a>
<a class="sourceLine" id="cb205-8" data-line-number="8"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#Number_of_cores_used&quot;</span>, NumbCoresToUse, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb205-9" data-line-number="9"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#MaxGlobalVariables&quot;</span>, MaxGlobalVariables, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb205-10" data-line-number="10"></a>
<a class="sourceLine" id="cb205-11" data-line-number="11">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Step&quot;</span>, <span class="st">&quot;Time(minutes)&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb205-12" data-line-number="12"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> OutfileCPUtimes, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb205-13" data-line-number="13"></a>
<a class="sourceLine" id="cb205-14" data-line-number="14"><span class="kw">lapply</span>(<span class="kw">names</span>(StopWatchStart), <span class="cf">function</span>(stepToClock) {</a>
<a class="sourceLine" id="cb205-15" data-line-number="15">  <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;POSIXct&quot;</span>, <span class="kw">class</span>(StopWatchStart[[stepToClock]]), <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb205-16" data-line-number="16">    TimeStart &lt;-<span class="st"> </span>StopWatchStart[[stepToClock]]</a>
<a class="sourceLine" id="cb205-17" data-line-number="17">    TimeEnd   &lt;-<span class="st"> </span>StopWatchEnd[[stepToClock]]</a>
<a class="sourceLine" id="cb205-18" data-line-number="18">    TimeDiff &lt;-<span class="st"> </span><span class="kw">format</span>(<span class="kw">difftime</span>(TimeEnd, TimeStart, <span class="dt">units =</span> <span class="st">&quot;min&quot;</span>))</a>
<a class="sourceLine" id="cb205-19" data-line-number="19">    ReportTime&lt;-<span class="kw">c</span>(<span class="kw">paste</span>(stepToClock, TimeDiff, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb205-20" data-line-number="20">    <span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x=</span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot; mins&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">x =</span> ReportTime), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb205-21" data-line-number="21">  }</a>
<a class="sourceLine" id="cb205-22" data-line-number="22">})</a></code></pre></div>
<pre><code>## [[1]]
## NULL</code></pre>
</div>
<div id="finish-2" class="section level3">
<h3><span class="header-section-number">1.0.24</span> FINISH</h3>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb207-2" data-line-number="2"><span class="co">### Finish</span></a>
<a class="sourceLine" id="cb207-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb207-4" data-line-number="4"><span class="kw">options</span>(<span class="dt">warn =</span> oldw)</a>
<a class="sourceLine" id="cb207-5" data-line-number="5"><span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">END - Check:</span><span class="ch">\n</span><span class="st">&quot;</span>, PathForOutfiles, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">For outfiles</span><span class="ch">\n\n</span><span class="st">&quot;</span>))</a></code></pre></div>
<pre><code>## 
## END - Check:
## /Users/jdiazmej/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/STACAS/WITHOUT_G945_I_T
## For outfiles</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb209-1" data-line-number="1"><span class="co"># quit()</span></a></code></pre></div>

<ul>
<li>Javier Diaz - <a href="mailto:javier.diazmejia@gmail.com" class="email">javier.diazmejia@gmail.com</a></li>
<li><p>Script based on: <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="uri">https://satijalab.org/seurat/articles/integration_introduction.html</a></p></li>
<li>GENERAL OVERVIEW OF THIS SCRIPT
<ol style="list-style-type: decimal">
<li>Loads integrated datasets R object produced by script CBW_CAN_2021_Module7_Lab3_PCA_Clustering_DimReduction.R</li>
<li>Computes differential gene expression (DGE) for each cell-class vs. the rest of cells</li>
<li>Computes DGE for one cell-class vs. another cell-class</li>
</ol></li>
</ul>
<hr />
</div>
<div id="define-environment-inputs-and-outdir-3" class="section level3">
<h3><span class="header-section-number">1.0.25</span> DEFINE ENVIRONMENT, INPUTS AND OUTDIR</h3>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb210-1" data-line-number="1">oldw &lt;-<span class="st"> </span><span class="kw">getOption</span>(<span class="st">&quot;warn&quot;</span>)</a>
<a class="sourceLine" id="cb210-2" data-line-number="2"><span class="kw">options</span>( <span class="dt">warn =</span> <span class="dv">-1</span> )</a></code></pre></div>
<div id="required-libraries-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.1</span> Required libraries</h6>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb211-1" data-line-number="1"><span class="kw">library</span>(Seurat)  <span class="co"># (CRAN) main scRNA-seq analysis package</span></a>
<a class="sourceLine" id="cb211-2" data-line-number="2"><span class="kw">library</span>(future)  <span class="co"># (CRAN) to run parallel processes</span></a></code></pre></div>
</div>
<div id="users-home-and-stopwatch-to-time-run-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.2</span> User’s home and stopwatch to time run</h6>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb212-1" data-line-number="1">UserHomeDirectory &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb212-2" data-line-number="2"><span class="kw">print</span>(UserHomeDirectory)</a></code></pre></div>
<pre><code>## [1] &quot;/Users/jdiazmej&quot;</code></pre>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb214-1" data-line-number="1">StopWatchStart &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb214-2" data-line-number="2">StopWatchEnd   &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb214-3" data-line-number="3"></a>
<a class="sourceLine" id="cb214-4" data-line-number="4">StopWatchStart<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a></code></pre></div>
</div>
<div id="define-inputs-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.3</span> Define inputs</h6>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb215-1" data-line-number="1"><span class="co">### Run specific inputs and parameters</span></a>
<a class="sourceLine" id="cb215-2" data-line-number="2"></a>
<a class="sourceLine" id="cb215-3" data-line-number="3">AnchorFinder        &lt;-<span class="st"> &quot;STACAS&quot;</span></a>
<a class="sourceLine" id="cb215-4" data-line-number="4">PathToIntegratedRds &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>, <span class="st">&quot;/Richards_NatCancer_2021_PCA_Clustering_DimReduction.rds&quot;</span>) <span class="co">### /path_to/*rds (R object) from CBW_CAN_2021_Module7_Lab3_PCA_Clustering_DimReduction.R</span></a>
<a class="sourceLine" id="cb215-5" data-line-number="5">InfileMetadata      &lt;-<span class="st"> &quot;~/CourseData/CAN_data/Module7/METADATA/Richards_NatCancer_2021_without_G945_I_T.metadata.tsv&quot;</span></a>
<a class="sourceLine" id="cb215-6" data-line-number="6">PathToIntegratedRds &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathToIntegratedRds)</a>
<a class="sourceLine" id="cb215-7" data-line-number="7">ASSAY               &lt;-<span class="st"> &quot;SCT&quot;</span> <span class="co">### Note, there is debate in the field about whether DGE should be calculated using RNA or SCT</span></a></code></pre></div>
</div>
<div id="define-outputs-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.4</span> Define outputs</h6>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb216-1" data-line-number="1"><span class="co">### Outputs</span></a>
<a class="sourceLine" id="cb216-2" data-line-number="2">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;~/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/&quot;</span>, AnchorFinder, <span class="st">&quot;/WITHOUT_G945_I_T&quot;</span>) <span class="co">## /path_to/out_directory</span></a>
<a class="sourceLine" id="cb216-3" data-line-number="3">PrefixOutfiles      &lt;-<span class="st"> &quot;Richards_NatCancer_2021&quot;</span>  <span class="co">## Prefix for outfiles</span></a>
<a class="sourceLine" id="cb216-4" data-line-number="4">PathForOutfiles     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;^~/&quot;</span>,<span class="kw">paste0</span>(UserHomeDirectory,<span class="st">&quot;/&quot;</span>), PathForOutfiles)</a>
<a class="sourceLine" id="cb216-5" data-line-number="5"><span class="kw">dir.create</span>(<span class="dt">path =</span> PathForOutfiles, <span class="dt">recursive =</span> T, <span class="dt">showWarnings =</span> F)</a></code></pre></div>
</div>
<div id="define-default-parameters-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.5</span> Define default parameters</h6>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb217-1" data-line-number="1"><span class="co">### Default parameters. Either suggested by Seurat developers, or tailored empirically.</span></a>
<a class="sourceLine" id="cb217-2" data-line-number="2">DefaultParameters &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb217-3" data-line-number="3">  </a>
<a class="sourceLine" id="cb217-4" data-line-number="4">  <span class="co">### Parameters for DGE calculation OneVsRest</span></a>
<a class="sourceLine" id="cb217-5" data-line-number="5">  <span class="dt">Test_OneVsRest               =</span> <span class="st">&quot;wilcox&quot;</span>,</a>
<a class="sourceLine" id="cb217-6" data-line-number="6">  <span class="dt">OnlyPos_OneVsRest            =</span> F,</a>
<a class="sourceLine" id="cb217-7" data-line-number="7">  <span class="dt">ReturnMinPctThresh_OneVsRest =</span> <span class="dv">0</span>,  <span class="co"># Use 0 if all cells should be included</span></a>
<a class="sourceLine" id="cb217-8" data-line-number="8">  <span class="dt">ReturnPvalThresh_OneVsRest   =</span> <span class="dv">1</span>,  <span class="co"># Use 1 if NO p-value cutoff should be used</span></a>
<a class="sourceLine" id="cb217-9" data-line-number="9">  <span class="dt">ReturnLogFcThresh_OneVsRest  =</span> <span class="dv">0</span>,  <span class="co"># Use 0 if NO LogFc cutoff should be used</span></a>
<a class="sourceLine" id="cb217-10" data-line-number="10"></a>
<a class="sourceLine" id="cb217-11" data-line-number="11">  <span class="co">### Parameters for DGE calculation Paired</span></a>
<a class="sourceLine" id="cb217-12" data-line-number="12">  <span class="dt">Test_Paired                  =</span> <span class="st">&quot;wilcox&quot;</span>,</a>
<a class="sourceLine" id="cb217-13" data-line-number="13">  <span class="dt">OnlyPos_Paired               =</span> F,</a>
<a class="sourceLine" id="cb217-14" data-line-number="14">  <span class="dt">ReturnMinPctThresh_Paired    =</span> <span class="dv">0</span>,  <span class="co"># Use 0 if all cells should be included</span></a>
<a class="sourceLine" id="cb217-15" data-line-number="15">  <span class="dt">ReturnPvalThresh_Paired      =</span> <span class="dv">1</span>,  <span class="co"># Use 1 if NO p-value cutoff should be used</span></a>
<a class="sourceLine" id="cb217-16" data-line-number="16">  <span class="dt">ReturnLogFcThresh_Paired     =</span> <span class="dv">0</span>,  <span class="co"># Use 0 if NO LogFc cutoff should be used</span></a>
<a class="sourceLine" id="cb217-17" data-line-number="17">  <span class="dt">subclass1_Paired             =</span> <span class="st">&quot;malignant&quot;</span>,</a>
<a class="sourceLine" id="cb217-18" data-line-number="18">  <span class="dt">subclass2_Paired             =</span> <span class="st">&quot;oligodendrocyte&quot;</span></a>
<a class="sourceLine" id="cb217-19" data-line-number="19">)</a></code></pre></div>
</div>
<div id="report-r-sessioninfo-3" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.6</span> Report R sessionInfo</h6>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb218-1" data-line-number="1">OutfileRSessionInfo&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_SingleCell_4_DGE_RSessionInfo.txt&quot;</span>)</a>
<a class="sourceLine" id="cb218-2" data-line-number="2"><span class="kw">writeLines</span>(<span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>()), OutfileRSessionInfo)</a>
<a class="sourceLine" id="cb218-3" data-line-number="3"><span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>())</a></code></pre></div>
<pre><code>##  [1] &quot;R version 4.0.2 (2020-06-22)&quot;                                                                                           
##  [2] &quot;Platform: x86_64-apple-darwin17.0 (64-bit)&quot;                                                                             
##  [3] &quot;Running under: macOS High Sierra 10.13.6&quot;                                                                               
##  [4] &quot;&quot;                                                                                                                       
##  [5] &quot;Matrix products: default&quot;                                                                                               
##  [6] &quot;BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib&quot;
##  [7] &quot;LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib&quot;                                    
##  [8] &quot;&quot;                                                                                                                       
##  [9] &quot;locale:&quot;                                                                                                                
## [10] &quot;[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8&quot;                                                      
## [11] &quot;&quot;                                                                                                                       
## [12] &quot;attached base packages:&quot;                                                                                                
## [13] &quot;[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     &quot;                          
## [14] &quot;&quot;                                                                                                                       
## [15] &quot;other attached packages:&quot;                                                                                               
## [16] &quot; [1] DropletUtils_1.8.0          SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.2&quot;                               
## [17] &quot; [4] DelayedArray_0.14.1         matrixStats_0.59.0          Biobase_2.48.0             &quot;                               
## [18] &quot; [7] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2         IRanges_2.22.2             &quot;                               
## [19] &quot;[10] S4Vectors_0.26.1            BiocGenerics_0.34.0         ggplot2_3.3.3              &quot;                               
## [20] &quot;[13] STACAS_1.1.0                future_1.21.0               SeuratObject_4.0.1         &quot;                               
## [21] &quot;[16] Seurat_4.0.2               &quot;                                                                                       
## [22] &quot;&quot;                                                                                                                       
## [23] &quot;loaded via a namespace (and not attached):&quot;                                                                             
## [24] &quot;  [1] plyr_1.8.6             igraph_1.2.6           lazyeval_0.2.2         splines_4.0.2         &quot;                      
## [25] &quot;  [5] BiocParallel_1.22.0    listenv_0.8.0          scattermore_0.7        digest_0.6.27         &quot;                      
## [26] &quot;  [9] htmltools_0.5.1.1      fansi_0.5.0            magrittr_2.0.1         tensor_1.5            &quot;                      
## [27] &quot; [13] cluster_2.1.2          ROCR_1.0-11            limma_3.44.3           globals_0.14.0        &quot;                      
## [28] &quot; [17] R.utils_2.10.1         spatstat.sparse_2.0-0  colorspace_2.0-1       ggrepel_0.9.1         &quot;                      
## [29] &quot; [21] xfun_0.23              dplyr_1.0.6            crayon_1.4.1           RCurl_1.98-1.3        &quot;                      
## [30] &quot; [25] jsonlite_1.7.2         spatstat.data_2.1-0    survival_3.2-11        zoo_1.8-9             &quot;                      
## [31] &quot; [29] glue_1.4.2             polyclip_1.10-0        gtable_0.3.0           zlibbioc_1.34.0       &quot;                      
## [32] &quot; [33] XVector_0.28.0         leiden_0.3.8           Rhdf5lib_1.10.1        future.apply_1.7.0    &quot;                      
## [33] &quot; [37] HDF5Array_1.16.1       abind_1.4-5            scales_1.1.1           edgeR_3.30.3          &quot;                      
## [34] &quot; [41] DBI_1.1.1              miniUI_0.1.1.1         Rcpp_1.0.6             viridisLite_0.4.0     &quot;                      
## [35] &quot; [45] xtable_1.8-4           dqrng_0.3.0            reticulate_1.20        spatstat.core_2.1-2   &quot;                      
## [36] &quot; [49] htmlwidgets_1.5.3      httr_1.4.2             RColorBrewer_1.1-2     ellipsis_0.3.2        &quot;                      
## [37] &quot; [53] ica_1.0-2              R.methodsS3_1.8.1      pkgconfig_2.0.3        farver_2.1.0          &quot;                      
## [38] &quot; [57] sass_0.4.0             uwot_0.1.10            deldir_0.2-10          locfit_1.5-9.4        &quot;                      
## [39] &quot; [61] utf8_1.2.1             tidyselect_1.1.1       labeling_0.4.2         rlang_0.4.11          &quot;                      
## [40] &quot; [65] reshape2_1.4.4         later_1.2.0            munsell_0.5.0          tools_4.0.2           &quot;                      
## [41] &quot; [69] generics_0.1.0         ggridges_0.5.3         evaluate_0.14          stringr_1.4.0         &quot;                      
## [42] &quot; [73] fastmap_1.1.0          yaml_2.2.1             goftest_1.2-2          knitr_1.33            &quot;                      
## [43] &quot; [77] fitdistrplus_1.1-5     purrr_0.3.4            RANN_2.6.1             packrat_0.6.0         &quot;                      
## [44] &quot; [81] pbapply_1.4-3          nlme_3.1-152           mime_0.10              R.oo_1.24.0           &quot;                      
## [45] &quot; [85] compiler_4.0.2         rstudioapi_0.13        plotly_4.9.3           png_0.1-7             &quot;                      
## [46] &quot; [89] spatstat.utils_2.1-0   tibble_3.1.2           bslib_0.2.5.1          stringi_1.6.2         &quot;                      
## [47] &quot; [93] highr_0.9              RSpectra_0.16-0        lattice_0.20-44        Matrix_1.3-3          &quot;                      
## [48] &quot; [97] vctrs_0.3.8            pillar_1.6.1           lifecycle_1.0.0        spatstat.geom_2.1-0   &quot;                      
## [49] &quot;[101] lmtest_0.9-38          jquerylib_0.1.4        RcppAnnoy_0.0.18       data.table_1.14.0     &quot;                      
## [50] &quot;[105] cowplot_1.1.1          bitops_1.0-7           irlba_2.3.3            httpuv_1.6.1          &quot;                      
## [51] &quot;[109] patchwork_1.1.1        R6_2.5.0               bookdown_0.22          promises_1.2.0.1      &quot;                      
## [52] &quot;[113] KernSmooth_2.23-20     gridExtra_2.3          parallelly_1.25.0      codetools_0.2-18      &quot;                      
## [53] &quot;[117] MASS_7.3-54            assertthat_0.2.1       rhdf5_2.32.4           withr_2.4.2           &quot;                      
## [54] &quot;[121] sctransform_0.3.2      GenomeInfoDbData_1.2.3 mgcv_1.8-35            grid_4.0.2            &quot;                      
## [55] &quot;[125] rpart_4.1-15           tidyr_1.1.3            rmarkdown_2.8          Rtsne_0.15            &quot;                      
## [56] &quot;[129] shiny_1.6.0           &quot;</code></pre>
</div>
<div id="define-number-of-cores-and-ram-for-parallelization-2" class="section level6">
<h6><span class="header-section-number">1.0.25.0.0.7</span> Define number of cores and RAM for parallelization</h6>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb220-1" data-line-number="1">NumbCores &lt;-<span class="st"> &quot;MAX&quot;</span></a>
<a class="sourceLine" id="cb220-2" data-line-number="2">MaxGlobalVariables &lt;-<span class="st"> </span><span class="dv">30000</span></a>
<a class="sourceLine" id="cb220-3" data-line-number="3"></a>
<a class="sourceLine" id="cb220-4" data-line-number="4"><span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^MAX$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb220-5" data-line-number="5">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">availableCores</span>()[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb220-6" data-line-number="6">}<span class="cf">else</span> <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;^[0-9]+$&quot;</span>, NumbCores, <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb220-7" data-line-number="7">  NumbCoresToUse &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(NumbCores)</a>
<a class="sourceLine" id="cb220-8" data-line-number="8">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb220-9" data-line-number="9">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected format for NumbCores: &quot;</span>, NumbCores, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb220-10" data-line-number="10">}</a>
<a class="sourceLine" id="cb220-11" data-line-number="11"></a>
<a class="sourceLine" id="cb220-12" data-line-number="12"><span class="cf">if</span> (NumbCoresToUse <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb220-13" data-line-number="13">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb220-14" data-line-number="14">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;sequential&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; core ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb220-15" data-line-number="15">}<span class="cf">else</span> <span class="cf">if</span> (NumbCoresToUse <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb220-16" data-line-number="16">  <span class="kw">plan</span>(<span class="dt">strategy =</span> <span class="st">&quot;multicore&quot;</span>, <span class="dt">workers =</span> NumbCoresToUse)</a>
<a class="sourceLine" id="cb220-17" data-line-number="17">  <span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="st">&quot;*** Running: in &#39;multicore&#39; mode with &quot;</span>, NumbCoresToUse, <span class="st">&quot; cores ***&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>))</a>
<a class="sourceLine" id="cb220-18" data-line-number="18">}<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb220-19" data-line-number="19">  <span class="kw">stop</span>(<span class="kw">paste0</span>(<span class="st">&quot;Unexpected NumbCores = &quot;</span>, NumbCoresToUse))</a>
<a class="sourceLine" id="cb220-20" data-line-number="20">}</a></code></pre></div>
<pre><code>## 
## *** Running: in &#39;multicore&#39; mode with 4 cores ***</code></pre>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb222-1" data-line-number="1"><span class="kw">options</span>(<span class="dt">future.globals.maxSize =</span> MaxGlobalVariables <span class="op">*</span><span class="st"> </span><span class="dv">1024</span><span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
</div>
</div>
<div id="load-r-object-1" class="section level3">
<h3><span class="header-section-number">1.0.26</span> LOAD R OBJECT</h3>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb223-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb223-2" data-line-number="2"><span class="co">### Load R Object</span></a>
<a class="sourceLine" id="cb223-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb223-4" data-line-number="4">seurat.object.integrated &lt;-<span class="st"> </span><span class="kw">readRDS</span>(PathToIntegratedRds)</a></code></pre></div>
</div>
<div id="load-metadata" class="section level3">
<h3><span class="header-section-number">1.0.27</span> LOAD METADATA</h3>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb224-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb224-2" data-line-number="2"><span class="co">### Loading metadata from --infile_metadata</span></a>
<a class="sourceLine" id="cb224-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb224-4" data-line-number="4">CellPropertiesFromMetadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">read.table</span>(InfileMetadata, <span class="dt">header =</span> T, <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">check.names =</span> F))</a>
<a class="sourceLine" id="cb224-5" data-line-number="5"><span class="kw">head</span>(CellPropertiesFromMetadata)</a></code></pre></div>
<pre><code>##                            cell_cluster cell_cluster.cell_type cell_type found_in_cell_cluster
## G1003_A_T_AAACCTGAGTATCGAA            1           01.malignant malignant                     1
## G1003_A_T_AAACCTGCAGTCCTTC            9           09.malignant malignant                     1
## G1003_A_T_AAACCTGTCAGTTTGG           11           11.malignant malignant                     1
## G1003_A_T_AAACCTGTCGGAATCT            2           02.malignant malignant                     1
## G1003_A_T_AAACCTGTCGTTTAGG            2           02.malignant malignant                     1
## G1003_A_T_AAACGGGAGAAACCGC            2           02.malignant malignant                     1
##                            dataset_id monochrome
## G1003_A_T_AAACCTGAGTATCGAA  G1003_A_T          1
## G1003_A_T_AAACCTGCAGTCCTTC  G1003_A_T          1
## G1003_A_T_AAACCTGTCAGTTTGG  G1003_A_T          1
## G1003_A_T_AAACCTGTCGGAATCT  G1003_A_T          1
## G1003_A_T_AAACCTGTCGTTTAGG  G1003_A_T          1
## G1003_A_T_AAACGGGAGAAACCGC  G1003_A_T          1</code></pre>
</div>
<div id="find-dge-using-global-cell-clusters-dimensions" class="section level3">
<h3><span class="header-section-number">1.0.28</span> FIND DGE USING GLOBAL CELL CLUSTERS DIMENSIONS</h3>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb226-1" data-line-number="1"><span class="kw">Idents</span>(<span class="dt">object =</span> seurat.object.integrated) &lt;-<span class="st"> &quot;seurat_clusters&quot;</span></a>
<a class="sourceLine" id="cb226-2" data-line-number="2"></a>
<a class="sourceLine" id="cb226-3" data-line-number="3"><span class="co">### For details on this approach to use a pseudocount, see https://f1000research.com/articles/7-1522</span></a>
<a class="sourceLine" id="cb226-4" data-line-number="4">FindMarkers.Pseudocount  &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw">length</span>(<span class="kw">rownames</span>(seurat.object.integrated<span class="op">@</span>meta.data))</a>
<a class="sourceLine" id="cb226-5" data-line-number="5"></a>
<a class="sourceLine" id="cb226-6" data-line-number="6"><span class="co">###  base = exp(1) is needed in Seurat v4 because the default uses Log2FC instead of LogNatFC, like v3 does</span></a>
<a class="sourceLine" id="cb226-7" data-line-number="7">seurat.object.integrated.markers &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> seurat.object.integrated,</a>
<a class="sourceLine" id="cb226-8" data-line-number="8">                                                   <span class="dt">assay =</span> ASSAY,</a>
<a class="sourceLine" id="cb226-9" data-line-number="9">                                                   <span class="dt">test.use =</span> DefaultParameters<span class="op">$</span>Test_OneVsRest,</a>
<a class="sourceLine" id="cb226-10" data-line-number="10">                                                   <span class="dt">only.pos =</span> DefaultParameters<span class="op">$</span>OnlyPos_OneVsRest,</a>
<a class="sourceLine" id="cb226-11" data-line-number="11">                                                   <span class="dt">min.pct =</span>  DefaultParameters<span class="op">$</span>ReturnMinPctThresh_OneVsRest,</a>
<a class="sourceLine" id="cb226-12" data-line-number="12">                                                   <span class="dt">return.thresh =</span>   DefaultParameters<span class="op">$</span>ReturnPvalThresh_OneVsRest,</a>
<a class="sourceLine" id="cb226-13" data-line-number="13">                                                   <span class="dt">logfc.threshold =</span> DefaultParameters<span class="op">$</span>ReturnLogFcThresh_OneVsRest,</a>
<a class="sourceLine" id="cb226-14" data-line-number="14">                                                   <span class="dt">pseudocount.use =</span> FindMarkers.Pseudocount,</a>
<a class="sourceLine" id="cb226-15" data-line-number="15">                                                   <span class="dt">base =</span> <span class="kw">exp</span>(<span class="dv">1</span>))</a></code></pre></div>
<pre><code>## Calculating cluster 0</code></pre>
<pre><code>## Calculating cluster 1</code></pre>
<pre><code>## Calculating cluster 2</code></pre>
<pre><code>## Calculating cluster 3</code></pre>
<pre><code>## Calculating cluster 4</code></pre>
<pre><code>## Calculating cluster 5</code></pre>
<pre><code>## Calculating cluster 6</code></pre>
<pre><code>## Calculating cluster 7</code></pre>
<pre><code>## Calculating cluster 8</code></pre>
<pre><code>## Calculating cluster 9</code></pre>
<pre><code>## Calculating cluster 10</code></pre>
<pre><code>## Calculating cluster 11</code></pre>
<pre><code>## Calculating cluster 12</code></pre>
<pre><code>## Calculating cluster 13</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb241-1" data-line-number="1"><span class="co">### IMPORTANT: use column &#39;gene&#39; instead of rownames</span></a>
<a class="sourceLine" id="cb241-2" data-line-number="2"><span class="kw">head</span>(seurat.object.integrated.markers)</a></code></pre></div>
<pre><code>##        p_val avg_logFC pct.1 pct.2 p_val_adj cluster   gene
## IL1A       0  3.255033 0.213 0.013         0       0   IL1A
## CCL4       0  3.249802 0.955 0.308         0       0   CCL4
## CCL3L3     0  3.231801 0.922 0.180         0       0 CCL3L3
## INHBA      0  3.090476 0.275 0.024         0       0  INHBA
## CCL4L2     0  3.040385 0.851 0.179         0       0 CCL4L2
## CCL3       0  2.974216 0.987 0.407         0       0   CCL3</code></pre>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb243-1" data-line-number="1">SimplifiedDiffExprGenes.df &lt;-<span class="st"> </span>seurat.object.integrated.markers[,<span class="kw">c</span>(<span class="st">&quot;cluster&quot;</span>,<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;p_val&quot;</span>,<span class="st">&quot;p_val_adj&quot;</span>,<span class="st">&quot;avg_logFC&quot;</span>,<span class="st">&quot;pct.1&quot;</span>,<span class="st">&quot;pct.2&quot;</span>)]</a>
<a class="sourceLine" id="cb243-2" data-line-number="2"></a>
<a class="sourceLine" id="cb243-3" data-line-number="3"><span class="co">### Add -Log10(Pval) * sign of FC</span></a>
<a class="sourceLine" id="cb243-4" data-line-number="4">SimplifiedDiffExprGenes.df[[<span class="st">&quot;mLog10Pval_FCsign&quot;</span>]] &lt;-<span class="st"> </span>(<span class="kw">log10</span>(SimplifiedDiffExprGenes.df[[<span class="st">&quot;p_val&quot;</span>]])<span class="op">*-</span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span><span class="kw">sign</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;avg_logFC&quot;</span>])</a>
<a class="sourceLine" id="cb243-5" data-line-number="5"><span class="co">### Replace -Inf and Inf in mLog10Pval_FCsign</span></a>
<a class="sourceLine" id="cb243-6" data-line-number="6">SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>] <span class="op">==</span><span class="st">  </span><span class="ot">Inf</span>]  &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">min</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]))<span class="op">*-</span><span class="dv">1</span></a>
<a class="sourceLine" id="cb243-7" data-line-number="7">SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>] <span class="op">==</span><span class="st">  </span><span class="op">-</span><span class="ot">Inf</span>] &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">max</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]))<span class="op">*-</span><span class="dv">1</span></a>
<a class="sourceLine" id="cb243-8" data-line-number="8"><span class="kw">head</span>(SimplifiedDiffExprGenes.df)</a></code></pre></div>
<pre><code>##        cluster   gene p_val p_val_adj avg_logFC pct.1 pct.2 mLog10Pval_FCsign
## IL1A         0   IL1A     0         0  3.255033 0.213 0.013          307.2867
## CCL4         0   CCL4     0         0  3.249802 0.955 0.308          307.2867
## CCL3L3       0 CCL3L3     0         0  3.231801 0.922 0.180          307.2867
## INHBA        0  INHBA     0         0  3.090476 0.275 0.024          307.2867
## CCL4L2       0 CCL4L2     0         0  3.040385 0.851 0.179          307.2867
## CCL3         0   CCL3     0         0  2.974216 0.987 0.407          307.2867</code></pre>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb245-1" data-line-number="1">OutfileDGE &lt;-<span class="st"> </span><span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_DGE_GlobalClustering_&quot;</span>, ASSAY, <span class="st">&quot;_&quot;</span>, DefaultParameters<span class="op">$</span>Test_OneVsRest, <span class="st">&quot;.tsv.bz2&quot;</span>)</a>
<a class="sourceLine" id="cb245-2" data-line-number="2">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(OutfileDGE, <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb245-3" data-line-number="3"><span class="kw">write.table</span>(<span class="dt">x =</span> <span class="kw">data.frame</span>(SimplifiedDiffExprGenes.df), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb245-4" data-line-number="4"><span class="kw">close</span>(Outfile.con)</a></code></pre></div>
</div>
<div id="find-dge-using-paired-metadata-classes" class="section level3">
<h3><span class="header-section-number">1.0.29</span> FIND DGE USING PAIRED METADATA CLASSES</h3>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb246-1" data-line-number="1"><span class="kw">Idents</span>(seurat.object.integrated) &lt;-<span class="st"> &quot;cell_type&quot;</span></a>
<a class="sourceLine" id="cb246-2" data-line-number="2">FindMarkers.Pseudocount  &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw">length</span>(<span class="kw">rownames</span>(seurat.object.integrated<span class="op">@</span>meta.data))</a>
<a class="sourceLine" id="cb246-3" data-line-number="3"><span class="co">###  base = exp(1) is needed in Seurat v4 because the default uses Log2FC instead of LogNatFC, like v3 does</span></a>
<a class="sourceLine" id="cb246-4" data-line-number="4">seurat.object.integrated.markers &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">FindMarkers</span>(<span class="dt">object =</span> seurat.object.integrated,</a>
<a class="sourceLine" id="cb246-5" data-line-number="5">                                                           <span class="dt">assay =</span> ASSAY,</a>
<a class="sourceLine" id="cb246-6" data-line-number="6">                                                           <span class="dt">test.use =</span> DefaultParameters<span class="op">$</span>Test_Paired,</a>
<a class="sourceLine" id="cb246-7" data-line-number="7">                                                           <span class="dt">only.pos =</span> DefaultParameters<span class="op">$</span>OnlyPos_Paired,</a>
<a class="sourceLine" id="cb246-8" data-line-number="8">                                                           <span class="dt">ident.1  =</span> DefaultParameters<span class="op">$</span>subclass1_Paired,</a>
<a class="sourceLine" id="cb246-9" data-line-number="9">                                                           <span class="dt">ident.2  =</span> DefaultParameters<span class="op">$</span>subclass2_Paired,</a>
<a class="sourceLine" id="cb246-10" data-line-number="10">                                                           <span class="dt">min.pct  =</span> DefaultParameters<span class="op">$</span>ReturnMinPctThresh_Paired,</a>
<a class="sourceLine" id="cb246-11" data-line-number="11">                                                           <span class="dt">return.thresh   =</span> DefaultParameters<span class="op">$</span>ReturnPvalThresh_Paired,</a>
<a class="sourceLine" id="cb246-12" data-line-number="12">                                                           <span class="dt">logfc.threshold =</span> DefaultParameters<span class="op">$</span>ReturnLogFcThresh_Paired,</a>
<a class="sourceLine" id="cb246-13" data-line-number="13">                                                           <span class="dt">pseudocount.use =</span> FindMarkers.Pseudocount,</a>
<a class="sourceLine" id="cb246-14" data-line-number="14">                                                           <span class="dt">base =</span> <span class="kw">exp</span>(<span class="dv">1</span>)))</a>
<a class="sourceLine" id="cb246-15" data-line-number="15">seurat.object.integrated.markers<span class="op">$</span>class1 &lt;-<span class="st"> </span>DefaultParameters<span class="op">$</span>subclass1_Paired</a>
<a class="sourceLine" id="cb246-16" data-line-number="16">seurat.object.integrated.markers<span class="op">$</span>class2 &lt;-<span class="st"> </span>DefaultParameters<span class="op">$</span>subclass2_Paired</a>
<a class="sourceLine" id="cb246-17" data-line-number="17">seurat.object.integrated.markers<span class="op">$</span>gene   &lt;-<span class="st"> </span><span class="kw">rownames</span>(seurat.object.integrated.markers)</a>
<a class="sourceLine" id="cb246-18" data-line-number="18"></a>
<a class="sourceLine" id="cb246-19" data-line-number="19">SimplifiedDiffExprGenes.df &lt;-<span class="st"> </span>seurat.object.integrated.markers[,<span class="kw">c</span>(<span class="st">&quot;class1&quot;</span>,<span class="st">&quot;class2&quot;</span>,<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;p_val&quot;</span>,<span class="st">&quot;p_val_adj&quot;</span>,<span class="st">&quot;avg_logFC&quot;</span>,<span class="st">&quot;pct.1&quot;</span>,<span class="st">&quot;pct.2&quot;</span>)]</a>
<a class="sourceLine" id="cb246-20" data-line-number="20"></a>
<a class="sourceLine" id="cb246-21" data-line-number="21"><span class="co">### Add -Log10(Pval) * sign of FC</span></a>
<a class="sourceLine" id="cb246-22" data-line-number="22">SimplifiedDiffExprGenes.df[[<span class="st">&quot;mLog10Pval_FCsign&quot;</span>]] &lt;-<span class="st"> </span>(<span class="kw">log10</span>(SimplifiedDiffExprGenes.df[[<span class="st">&quot;p_val&quot;</span>]])<span class="op">*-</span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span><span class="kw">sign</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;avg_logFC&quot;</span>])</a>
<a class="sourceLine" id="cb246-23" data-line-number="23"><span class="co">### Replace -Inf and Inf in mLog10Pval_FCsign</span></a>
<a class="sourceLine" id="cb246-24" data-line-number="24">SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>] <span class="op">==</span><span class="st">  </span><span class="ot">Inf</span>]  &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">min</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]))<span class="op">*-</span><span class="dv">1</span></a>
<a class="sourceLine" id="cb246-25" data-line-number="25">SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;mLog10Pval_FCsign&quot;</span>] <span class="op">==</span><span class="st">  </span><span class="op">-</span><span class="ot">Inf</span>] &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">max</span>(SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>][SimplifiedDiffExprGenes.df[,<span class="st">&quot;p_val&quot;</span>] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]))<span class="op">*-</span><span class="dv">1</span></a>
<a class="sourceLine" id="cb246-26" data-line-number="26"><span class="kw">head</span>(SimplifiedDiffExprGenes.df)</a></code></pre></div>
<pre><code>##            class1          class2    gene p_val p_val_adj  avg_logFC pct.1 pct.2 mLog10Pval_FCsign
## VWA1    malignant oligodendrocyte    VWA1     0         0 -1.9362850 0.202 0.757            0.0000
## RBP7    malignant oligodendrocyte    RBP7     0         0 -5.0060315 0.005 0.510            0.0000
## CAMK2N1 malignant oligodendrocyte CAMK2N1     0         0 -2.4241825 0.289 0.930            0.0000
## RPL11   malignant oligodendrocyte   RPL11     0         0  0.9726906 0.998 0.986          307.2163
## YBX1    malignant oligodendrocyte    YBX1     0         0  1.5010289 0.991 0.873          307.2163
## TMEM125 malignant oligodendrocyte TMEM125     0         0 -5.6157419 0.005 0.637            0.0000</code></pre>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb248-1" data-line-number="1">OutfileDGE &lt;-<span class="st"> </span><span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_DGE_Paired_&quot;</span>, ASSAY, <span class="st">&quot;_&quot;</span>, DefaultParameters<span class="op">$</span>Test_Paired, <span class="st">&quot;.tsv.bz2&quot;</span>)</a>
<a class="sourceLine" id="cb248-2" data-line-number="2">Outfile.con &lt;-<span class="st"> </span><span class="kw">bzfile</span>(OutfileDGE, <span class="st">&quot;w&quot;</span>)</a>
<a class="sourceLine" id="cb248-3" data-line-number="3"><span class="kw">write.table</span>(<span class="dt">x =</span> <span class="kw">data.frame</span>(SimplifiedDiffExprGenes.df), <span class="dt">file =</span> Outfile.con, <span class="dt">row.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb248-4" data-line-number="4"><span class="kw">close</span>(Outfile.con)</a></code></pre></div>
</div>
<div id="obtain-computing-time-3" class="section level3">
<h3><span class="header-section-number">1.0.30</span> OBTAIN COMPUTING TIME</h3>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb249-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb249-2" data-line-number="2"><span class="co">### Obtain computing time used</span></a>
<a class="sourceLine" id="cb249-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb249-4" data-line-number="4">StopWatchEnd<span class="op">$</span>Overall  &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a>
<a class="sourceLine" id="cb249-5" data-line-number="5">OutfileCPUtimes&lt;-<span class="kw">paste0</span>(PathForOutfiles, <span class="st">&quot;/&quot;</span>, PrefixOutfiles, <span class="st">&quot;_DGE_CPUtimes.txt&quot;</span>)</a>
<a class="sourceLine" id="cb249-6" data-line-number="6"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#Number_of_cores_used&quot;</span>, NumbCoresToUse, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb249-7" data-line-number="7"><span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x =</span> <span class="kw">paste</span>(<span class="st">&quot;#MaxGlobalVariables&quot;</span>, MaxGlobalVariables, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb249-8" data-line-number="8"></a>
<a class="sourceLine" id="cb249-9" data-line-number="9">Headers&lt;-<span class="kw">paste</span>(<span class="st">&quot;Step&quot;</span>, <span class="st">&quot;Time(minutes)&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb249-10" data-line-number="10"><span class="kw">write.table</span>(Headers, <span class="dt">file =</span> OutfileCPUtimes, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> F, <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb249-11" data-line-number="11"></a>
<a class="sourceLine" id="cb249-12" data-line-number="12"><span class="kw">lapply</span>(<span class="kw">names</span>(StopWatchStart), <span class="cf">function</span>(stepToClock) {</a>
<a class="sourceLine" id="cb249-13" data-line-number="13">  <span class="cf">if</span> (<span class="kw">regexpr</span>(<span class="st">&quot;POSIXct&quot;</span>, <span class="kw">class</span>(StopWatchStart[[stepToClock]]), <span class="dt">ignore.case =</span> T)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb249-14" data-line-number="14">    TimeStart &lt;-<span class="st"> </span>StopWatchStart[[stepToClock]]</a>
<a class="sourceLine" id="cb249-15" data-line-number="15">    TimeEnd   &lt;-<span class="st"> </span>StopWatchEnd[[stepToClock]]</a>
<a class="sourceLine" id="cb249-16" data-line-number="16">    TimeDiff &lt;-<span class="st"> </span><span class="kw">format</span>(<span class="kw">difftime</span>(TimeEnd, TimeStart, <span class="dt">units =</span> <span class="st">&quot;min&quot;</span>))</a>
<a class="sourceLine" id="cb249-17" data-line-number="17">    ReportTime&lt;-<span class="kw">c</span>(<span class="kw">paste</span>(stepToClock, TimeDiff, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb249-18" data-line-number="18">    <span class="kw">write</span>(<span class="dt">file =</span> OutfileCPUtimes, <span class="dt">x=</span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot; mins&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">x =</span> ReportTime), <span class="dt">append =</span> T)</a>
<a class="sourceLine" id="cb249-19" data-line-number="19">  }</a>
<a class="sourceLine" id="cb249-20" data-line-number="20">})</a></code></pre></div>
<pre><code>## [[1]]
## NULL</code></pre>
</div>
<div id="finish-3" class="section level3">
<h3><span class="header-section-number">1.0.31</span> FINISH</h3>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb251-1" data-line-number="1"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb251-2" data-line-number="2"><span class="co">### Finish</span></a>
<a class="sourceLine" id="cb251-3" data-line-number="3"><span class="co">####################################</span></a>
<a class="sourceLine" id="cb251-4" data-line-number="4"><span class="kw">options</span>(<span class="dt">warn =</span> oldw)</a>
<a class="sourceLine" id="cb251-5" data-line-number="5"><span class="kw">writeLines</span>(<span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">END - Check:</span><span class="ch">\n</span><span class="st">&quot;</span>, PathForOutfiles, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">For outfiles</span><span class="ch">\n\n</span><span class="st">&quot;</span>))</a></code></pre></div>
<pre><code>## 
## END - Check:
## /Users/jdiazmej/CourseData/CAN_data/Module7/OUTFILES/Richards_NatCancer_2021/STACAS/WITHOUT_G945_I_T
## For outfiles</code></pre>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb253-1" data-line-number="1"><span class="co"># quit()</span></a></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/jdime/CBW_2021_CAN_M7/edit/master/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/jdime/CBW_2021_CAN_M/blob/master/index.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
